---
title: "colocalization analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ColocAnalysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette shows a complete workflow of colocalization analysis with your own GWAS data using GTExbiolinks.

```{r setup}
library(GTExbiolinks)
library(data.table)
library(utils)
```

Step 1: download example file of GWAS summary statistics data.
```{r}
gwasFile <- tempfile(pattern = "file")
gwasURL <- "http://bioinfo.szbl.ac.cn/finalColoc/tmp/gwasFile/gwasChr6Sub1.txt"
utils::download.file(gwasURL, destfile=gwasFile)
# If your dataset provides the standard error, simply square it to get the variance.
gwasDF <- data.table::fread(gwasFile, sep="\t", header=TRUE)
gwasDF
```

Step 2: fetch all regulatory associations of variants for gene of interest, like "RP11-385F7.1" in this example.
```{r fig.width=6.1, fig.height=4.7, message=FALSE}
traitGene="RP11-385F7.1"
tissueSiteDetail="Brain - Cortex"
geneAsso <- GTExvisual_eqtlTrait(traitGene, tissueSiteDetail=tissueSiteDetail)
```

Keep three columns for following analyses:
```{r}
geneAsso <- geneAsso[,.(snpId, pValue, maf)] 
geneAsso
```

Step 3: merge gwas and eqtl datasets by SNP ID. Centered on the SNP with the lowest pValue, sentinel snps within one million bases are retained.
```{r}
gwas_eqtl <- merge(gwasDF[,.(snpId=rsid, pValue=P, maf=maf, position)], geneAsso, by="snpId", sort=FALSE, suffixes = c(".gwas",".eqtl"))
centerSnp <- gwas_eqtl[which.min(gwas_eqtl$pValue.gwas),]
colocRange=1e6
sentinelSnps <- gwas_eqtl[position>=(centerSnp$position-(colocRange)/2) & position<=(centerSnp$position+(colocRange)/2),]
sentinelSnps
```

Step 4: conduct colocalization analysis using coloc package.
```{r}
library(coloc)
colocResult <- coloc.abf(
  dataset1=list(
    pvalues=sentinelSnps$pValue.gwas, type="quant", N=50000, MAF=sentinelSnps$maf.gwas),
 dataset2=list(
   pvalues=sentinelSnps$pValue.eqtl, type="quant", N=600,
   MAF=sentinelSnps$maf.gwas))
colocResult
```

Step 5: visualize GWAS-eQTL colocalization for selected gene. 
```{r fig.width=6.1, fig.height=4.7, message=FALSE}
eqtlAsso2 <-  GTExvisual_eqtlGWAS(sentinelSnps[,c("snpId","pValue.gwas")], traitGene="RP11-385F7.1", tissueSiteDetail="Lung")
```

Or you can use locuscomparer package for visualization of GWAS-eQTL colocalization.
```{r fig.width=6.1, fig.height=4.7, warning=FALSE}
if(require(locuscomparer)){
  locuscompare( sentinelSnps[,.(rsid=snpId, pval=pValue.gwas)], sentinelSnps[,.(rsid=snpId, pval=pValue.eqtl)], snp = "rs1931837" )
}
```

Step 5: effect of the variant "rs1931837" on expression of gene "RP11-385F7.1" in lung.
```{r  fig.width=6.1, fig.height=4.7, warning=FALSE, message=FALSE}
expEqtl <- GTExvisual_eqtlExp(variantName="rs1931837", gene ="RP11-385F7.1", tissueSiteDetail="Lung")
```


