---
title: "Colocalization_analysis_with_xQTLbiolinks"
date: "`r format(Sys.time(), '%Y-%b-%d')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Colocalization_analysis_with_xQTLbiolinks}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  echo=TRUE,
  progress =FALSE,
  collapse = TRUE,
  comment = ">"
)
```



## Before we start. 
All we need to prepare include two parts: 

  1. GWAS summary statistics dataset.
  2. Name of tissue of interest. 
  
In this example, we download summary statistics dataset of a GWAS study ([GCST006085](http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST006001-GCST007000/GCST006085/harmonised/29892016-GCST006085-EFO_0001663-build37.f.tsv.gz)) of prostate cancer from GWAS category and load the dataset in R with `data.table` package. Correspondingly, we chose tissue `Porstate` for study.

```{r setup, message=FALSE}
library(data.table)
library(xQTLbiolinks)
library(stringr)
```

We retain the variants with dbSNP id (start with `rs`), and a data.table object named `gwasDF` of 13,498,990 (rows) x 5 (cols) is loaded.

```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
gwasDF <- fread("D:\\R_project\\Data\\GLGC_CG0276_result.txt")
# extract columns.
gwasDF<- gwasDF[str_detect(variant_id, "^rs"),.(rsid=variant_id, chrom=chromosome, position= base_pair_location, pValue=p_value, AF=effect_allele_frequency)]
# tissue:
tissueSiteDetail="Prostate"
```

``` {r message=FALSE, warning=FALSE, include=FALSE}
gwasDF <- example_Coloc_gwasDF
```

```{r eval=TRUE}
head(gwasDF)
```

## Step 1: Filter sentinel snps. 

Sentinel SNP is the most prominent signal within a given genome range, and is usually in high LD with causal variants. By default, xQTLbiolinks detect sentinel snps that with the p-value < 5e-8 and SNP-to-SNP distance > 10e6 bp. **Note**: For in this example, due to the inconsistent genome version between the GWAS dataset (GRCh37) and eQTL associations (GRCh38) from eQTL category, conversion of genome version is required, and can be conducted using `xQTLanalyze_getSentinelSnp` with `genomeVersion="grch37"` and `grch37To38=TRUE`:

```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
sentinelSnpDF <- xQTLanalyze_getSentinelSnp(gwasDF, centerRange=1e6,
                                            genomeVersion="grch37", grch37To38=TRUE)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
sentinelSnpDF <- example_Coloc_sentinelSNP
```

A total of 94 sentinel SNPs are detected.
```{r eval=TRUE}
head(sentinelSnpDF)
```

## Step 2: Identify trait genes for each sentinel SNPs. 

Trait genes are genes that located in the range of 1Mb (default, can be changed with parameter `detectRange`) of sentinel SNPs. In order to reduce the number of trait genes and thus reduce the running time, we take the overlap of eGenes and trait genes as the final output of the function `xQTLanalyze_getTraits`:

```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
traitsAll <- xQTLanalyze_getTraits(sentinelSnpDF, detectRange=1e6, tissueSiteDetail=tissueSiteDetail)
```

Totally, 898 associations between 835 traits genes and 92 sentinel SNPs are detected

```{r message=FALSE, warning=FALSE, include=FALSE}
tissueSiteDetail="Prostate"
traitsAll <- example_Coloc_traitsAll
```

```{r eval=TRUE}
head(example_Coloc_traitsAll)
```


## Step 3: Conduct colocalization analysis. 

Following three steps of colocalization analysis are encapsulated in one function xQTLanalyze_coloc:

1. Retrieved all associations from EBI eQTL catalogue for a specified gene.
2. Merge the data.frame of GWAS and eQTL by rsid.
3. Perform colocalization analysis usingcoloc package.


For above 835 trait genes, a for loop can be used to get these genes' outputs of colocalization analysis (this may take several hours):
```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
genesAll<- unique(traitsAll$gencodeId)
colocResultAll <- data.table()
for(i in 1:length(genesAll)){
  colocResult <- xQTLanalyze_coloc(gwasDF, genomeVersion = "grch37", genesAll[i], tissueSiteDetail=tissueSiteDetail)$coloc_Out_summary
  if(!is.null(colocResult)){ colocResultAll <- rbind(colocResultAll, colocResult)}
  message(format(Sys.time(), "== %Y-%b-%d %H:%M:%S ")," == Id:",i,"/",length(genesAll)," == Gene:",genesAll[i], " == PP4: ", colocResult$PP.H4.abf)
}
```
  
```{r message=FALSE, warning=FALSE, include=FALSE}
colocResultAll <- example_Coloc_colocResultAll
```

In this case, we invoke the funciton `coloc.abf` from package `coloc` to estimate the posterior support of variants for each hypothesis: H0,H1,H2,H3,H4:
  
- H0 : neither trait has a genetic association in the region
- H1: only trait 1 has a genetic association in the region
- H2: only trait 2 has a genetic association in the region
- H3: both traits are associated, but with different causal variants
- H4: both traits are associated and share a single causal variant

Output is a data.table that combined all results of `coloc_Out_summary` of all genes. The posterior probability of variants fo each hypothesis (H0-H4) are listed in `colocResultAll`:
```{r eval=TRUE}
head(colocResultAll)
```

## Step 4: Visualization of the results.

We considered colocalization tests with a posterior probability of hypothesis 4 (PPH4.ABF) â‰¥ 0.75 as having strong or moderate evidence for colocalization.
  
```{r message=FALSE, warning=FALSE, include=FALSE}
colocResultsig <- example_Coloc_colocResultsig
```

```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
colocResultsig <- colocResultAll[PP.H4.abf>0.75][order(-PP.H4.abf)]
```
There are 27 trait genes that are associated and share a single causal variant:
```{r eval=TRUE}
head(colocResultsig)
```
For the potential cancer-related trait gene `ENSG00000137673.8` that with the highest `PPH4.ABF=0.9978`, gene's detail can be fetched:
```{r eval=TRUE, message=FALSE, warning=FALSE}
geneInfo <- xQTLquery_gene("ENSG00000137673.8")
geneInfo
```
  
  
```{r message=FALSE, warning=FALSE, include=FALSE}
t1 <- tempfile()
utils::download.file("https://gitee.com/stronghoney/exampleData/raw/master/gwasEqtldata.txt", t1)
gwasEqtldata <- data.table::fread(t1)
```

Merge the variants of GWAS and eQTL dataset by rsid:
```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
eqtlAsso <- xQTLdownload_eqtlAllAsso(gene="ENSG00000137673.8",tissueSiteDetail = tissueSiteDetail, 
                                     withB37VariantId=FALSE)
# Here we use genome position of hg19 to visualization:
gwasEqtldata <- merge(gwasDF[,-c("AF")], eqtlAsso[,.(rsid=snpId, pValue)],
                      by=c("rsid"), suffixes = c(".gwas",".eqtl"))
```

Five fields are retained:
```{r eval=TRUE}
gwasEqtldata
```
  
Visualization of P-value distribution and comparison of the signals of GWAS and eQTL:
```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
xQTLvisual_locusCompare(gwasEqtldata[,.(rsid, pValue.eqtl)], 
                        gwasEqtldata[,.(rsid, pValue.gwas)], legend_position = "bottomright")
```
```{r, fig.width=6, fig.height=5.5, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("prostate_locuscompare.png")
grid.raster(img)
```
  
Locuszoom plot of GWAS signals:
```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
xQTLvisual_locusZoom(gwasEqtldata[,.(rsid, chrom, position, pValue.gwas)], legend=FALSE)
```
```{r, fig.width=6, fig.height=5, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("prostate_locuszoom_gwas.png")
grid.raster(img)
```
  
Locuszoom plot of eQTL signals:
```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
xQTLvisual_locusZoom(gwasEqtldata[,.(rsid, chrom, position, pValue.eqtl)], legend=FALSE)
```
```{r, fig.width=6, fig.height=5, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("prostate_locuszoom_eqtl.png")
grid.raster(img)
```
  
We can also combine locuscompare and locuszoom plot using function `xQTLvisual_locusCombine`:
```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
xQTLvisual_locusCombine(gwasEqtldata[,c("rsid","chrom", "position", "pValue.gwas", "pValue.eqtl")])
```
```{r, fig.width=6, fig.height=5, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("prostate_combined.png")
grid.raster(img)
```
  
From the above figures, we can see that the SNP `rs11568818` is potential causal variant, and we can use a violin plot to show the normalized effect size of it:
```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
xQTLvisual_eqtlExp("rs11568818", "ENSG00000137673.8", tissueSiteDetail = tissueSiteDetail)
```
```{r, fig.width=3.5, fig.height=3.2, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("prostate_eQTL_exp.png")
grid.raster(img)
```
  
visualization of the gene expression among multiple tissues:
```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
geneExpTissues <- xQTLvisual_geneExpTissues("ENSG00000137673.8", tissues="All")
```


## Extension: Integrative analysis with external packages or tools

To gain insight into the function of trait genes with PPH4>0.75, and explore the potential regulatory mechanism of the prostate cancer, we first fetched all these genes' details with `xQTLquery_gene`:

```{r eval=TRUE, message=FALSE}
library(SummarizedExperiment)

colocResultsigGene <- xQTLquery_gene(colocResultsig$traitGene)
```
Add the value of PPH4 for each gene:
```{r eval=TRUE}
colocResultsigGene <- merge(colocResultsig[,.(gencodeId= traitGene, PP.H4.abf)], 
                            colocResultsigGene[,.(geneSymbol, gencodeId, entrezGeneId, geneType)], by="gencodeId", sort=FALSE)
head(colocResultsigGene)
```

To explore the genes that co-expressed with above genes, we download expression profiles of all protein-coding genes in prostate.

```{r results = 'hide', echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
prot <- xQTLquery_gene(genes="protein coding")
protExp <- xQTLdownload_exp(prot$gencodeId, tissueSiteDetail=tissueSiteDetail)
# Extract expression profiles: 
expMat <- SummarizedExperiment::assay(protExp)
```

```{r eval=TRUE}
expZip <- tempfile(fileext = ".zip")
utils::download.file("https://raw.githubusercontent.com/dingruofan/exampleData/master/expmat.zip", expZip)
expMat<- read.table(unzip(expZip, "expmat.txt", exdir = dirname(expZip)), row.names = 1, fill=TRUE, header = TRUE, check.names = FALSE)
```

```{r eval=TRUE}
expMat[1:5,1:4]
```

We can calculate pearson coefficient with the expression matrix for each trait gene:
```
<!-- for( i in 1:nrow(colocResultsigGene)){ -->
<!--   lapply(colocResultsigGene[i]$gencodeId) -->
<!-- } -->
```



