<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>3. Case study: colocalization analysis in prostate cancer • xQTLbiolinks</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="3. Case study: colocalization analysis in prostate cancer">
<meta property="og:description" content="xQTLbiolinks">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xQTLbiolinks</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/visualization.html">Visualization</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dingruofan/xQTLbiolinks/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>3. Case study: colocalization analysis in
prostate cancer</h1>
            
            <h4 data-toc-skip class="date">2023-05-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dingruofan/xQTLbiolinks/blob/HEAD/vignettes/Colocalization_analysis_with_xQTLbiolinks.Rmd" class="external-link"><code>vignettes/Colocalization_analysis_with_xQTLbiolinks.Rmd</code></a></small>
      <div class="hidden name"><code>Colocalization_analysis_with_xQTLbiolinks.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 id="note-the-gtex-portal-api-update-has-caused-some-functions-of-xqtlbioliks-to-malfunction--the-package-will-be-updated-within-one-week-to-address-this-issue-">
<strong>NOTE! The GTEx Portal API update has caused some functions
of xQTLbioliks to malfunction. The package will be updated within one
week to address this issue.</strong><a class="anchor" aria-label="anchor" href="#note-the-gtex-portal-api-update-has-caused-some-functions-of-xqtlbioliks-to-malfunction--the-package-will-be-updated-within-one-week-to-address-this-issue-"></a>
</h2>
<hr>
</div>
<div class="section level2">
<h2 id="before-we-start-">Before we start.<a class="anchor" aria-label="anchor" href="#before-we-start-"></a>
</h2>
<p>All we need to prepare include three parts:</p>
<ol style="list-style-type: decimal">
<li>GWAS summary statistics dataset.</li>
<li>Tissue of interest.</li>
<li>Following three installed packages.</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dingruofan/xQTLbiolinks" class="external-link">xQTLbiolinks</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chr1swallace/coloc" class="external-link">coloc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hyprcoloc</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://henrikbengtsson.github.io/R.utils/" class="external-link">R.utils</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span></span></code></pre></div>
<p>Prostate cancer is one of the most common cancers in men. Prostate
cancer pathogenesis involves both heritable and environmental factors.
The molecular events involved in the development or progression of
prostate cancer are still unclear. In this example, we aim to figure out
the causal variants and genes assocaited with prostate cancer, and to
uncover potential molecular mechanisms of regulation.</p>
<p>For data preparation, we download summary statistics dataset of a
GWAS study (<a href="http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST006001-GCST007000/GCST006085/harmonised/29892016-GCST006085-EFO_0001663-build37.f.tsv.gz" class="external-link">GCST006085</a>)
of prostate cancer from GWAS category and load the dataset in R with
<code>data.table</code> package. Correspondingly, we chose tissue
<code>Prostate</code> for study. We retain the variants with dbSNP id
(start with <code>rs</code>), and a data.table object named
<code>gwasDF</code> of 13,498,990 (rows) x 7 (cols) is loaded.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwasDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"29892016-GCST006085-EFO_0001663-build37.f.tsv.gz"</span><span class="op">)</span></span>
<span><span class="co"># extract columns.</span></span>
<span><span class="va">gwasDF</span><span class="op">&lt;-</span> <span class="va">gwasDF</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">variant_id</span>, <span class="st">"^rs"</span><span class="op">)</span>,<span class="fu">.</span><span class="op">(</span>rsid<span class="op">=</span><span class="va">variant_id</span>, chrom<span class="op">=</span><span class="va">chromosome</span>, position<span class="op">=</span> <span class="va">base_pair_location</span>, pValue<span class="op">=</span><span class="va">p_value</span>, AF<span class="op">=</span><span class="va">effect_allele_frequency</span>, <span class="va">beta</span>, se<span class="op">=</span> <span class="va">standard_error</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># tissue:</span></span>
<span><span class="va">tissueSiteDetail</span><span class="op">=</span><span class="st">"Prostate"</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;           rsid chrom position  pValue     AF    beta     se</span></span>
<span><span class="co">#&gt; 1: rs145072688     1    10352 0.46280 0.5753 -0.0094 0.0128</span></span>
<span><span class="co">#&gt; 2: rs376342519     1    10616 0.41070 0.0066  0.0572 0.0696</span></span>
<span><span class="co">#&gt; 3: rs201725126     1    13116 0.84470 0.8255 -0.0035 0.0178</span></span>
<span><span class="co">#&gt; 4: rs200579949     1    13118 0.84470 0.8255 -0.0035 0.0178</span></span>
<span><span class="co">#&gt; 5:  rs75454623     1    14930 0.61880 0.4772 -0.0065 0.0131</span></span>
<span><span class="co">#&gt; 6: rs199856693     1    14933 0.02721 0.0407 -0.0794 0.0360</span></span></code></pre>
<p>Now we can perform quality control for this GWAS dataset by
estimating the genomic inflation factor, validating QQ-plot and PZ-plot
using xQTLbiolinks.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_qqPlot.html">xQTLvisual_qqPlot</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">pValue</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/xQTLvisual_PZPlot.html">xQTLvisual_PZPlot</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">pValue</span>, <span class="va">beta</span>, <span class="va">se</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/qq_pz_plot.png" width="80%" style="display: block; margin: auto;"></p>
<p>Left is the QQ-plot. The x-axis of the QQ-plot represents the
expected p-values under the null hypothesis, while the y-axis represents
the observed p-values from the GWAS summary statistics data. There is a
significant deviation from the diagonal line, which indicates potential
variations from the null hypothesis that may result from true
associations or LD. The genomic inflation factor (lambda) is labeled in
the top left of the QQ-plot, and a lambda value of 1.138 indicating no
strong population stratification. Right is the PZ plot, the observed
p-values on the y-axis and the corresponding p-values derived from
z-scores that estimated by beta and se on the x-axis. A strong
concordance can be observed between the observed p-values and those
calculated from Z-scores.</p>
<p>Genomic annotation of the significant signals can be conducted using
<code>xQTLanno_genomic</code>:</p>
</div>
<div class="section level2">
<h2 id="step-1-identify-sentinel-snps">Step 1: Identify sentinel SNPs:<a class="anchor" aria-label="anchor" href="#step-1-identify-sentinel-snps"></a>
</h2>
<p>Sentinel SNP is the most prominent signal within a given genome
range, and is usually in high LD with causal variants. By default,
xQTLbiolinks detect sentinel snps that with the p-value &lt; 5e-8 and
SNP-to-SNP distance &gt; 10e6 bp. <strong>Note</strong>: For in this
example, due to the inconsistent genome version between the GWAS dataset
(GRCh37) and eQTL associations (GRCh38) from eQTL category, conversion
of genome version is required, and can be conducted using
<code>xQTLanalyze_getSentinelSnp</code> with
<code>genomeVersion="grch37"</code> and <code>grch37To38=TRUE</code>
(package <code>rtracklayer</code>is required):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sentinelSnpDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_getSentinelSnp.html">xQTLanalyze_getSentinelSnp</a></span><span class="op">(</span><span class="va">gwasDF</span>, centerRange<span class="op">=</span><span class="fl">1e6</span>,</span>
<span>                                            genomeVersion<span class="op">=</span><span class="st">"grch37"</span>, grch37To38<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>A total of 94 sentinel SNPs are detected.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sentinelSnpDF</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;           rsid  chr  position    pValue    maf    beta     se</span></span>
<span><span class="co">#&gt; 1:  rs55664108 chr1 204587862 8.572e-25 0.7157  0.0912 0.0089</span></span>
<span><span class="co">#&gt; 2:  rs35296356 chr1 150601662 3.355e-14 0.3384 -0.0690 0.0091</span></span>
<span><span class="co">#&gt; 3:  rs34579442 chr1 153927424 4.478e-14 0.6637 -0.0655 0.0087</span></span>
<span><span class="co">#&gt; 4: rs146564277 chr1 155051619 1.577e-12 0.0301 -0.1850 0.0262</span></span>
<span><span class="co">#&gt; 5:  rs34848415 chr1 205762038 2.893e-09 0.5231 -0.0501 0.0084</span></span>
<span><span class="co">#&gt; 6:  rs56391074 chr1  87745032 1.659e-08 0.6298 -0.0466 0.0082</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-2-identify-trait-genes-for-each-sentinel-snps">Step 2: Identify trait genes for each sentinel SNPs:<a class="anchor" aria-label="anchor" href="#step-2-identify-trait-genes-for-each-sentinel-snps"></a>
</h2>
<p>Trait genes are genes that located in the range of 1Mb (default, can
be changed with parameter <code>detectRange</code>) of sentinel SNPs. In
order to reduce the number of trait genes and thus reduce the running
time, we take the overlap of eGenes and trait genes as the final output
of the function <code>xQTLanalyze_getTraits</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traitsAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_getTraits.html">xQTLanalyze_getTraits</a></span><span class="op">(</span><span class="va">sentinelSnpDF</span>, detectRange<span class="op">=</span><span class="fl">1e6</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span></span></code></pre></div>
<p>Totally, 898 associations between 835 traits genes and 92 sentinel
SNPs are detected</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">traitsAll</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;          rsid  chr  position    pValue    maf    beta     se       gencodeId</span></span>
<span><span class="co">#&gt; 1: rs34848415 chr1 205762038 2.893e-09 0.5231 -0.0501 0.0084 ENSG00000261000</span></span>
<span><span class="co">#&gt; 2: rs34848415 chr1 205762038 2.893e-09 0.5231 -0.0501 0.0084 ENSG00000263528</span></span>
<span><span class="co">#&gt; 3: rs34848415 chr1 205762038 2.893e-09 0.5231 -0.0501 0.0084 ENSG00000162877</span></span>
<span><span class="co">#&gt; 4: rs34848415 chr1 205762038 2.893e-09 0.5231 -0.0501 0.0084 ENSG00000117280</span></span>
<span><span class="co">#&gt; 5: rs34848415 chr1 205762038 2.893e-09 0.5231 -0.0501 0.0084 ENSG00000069275</span></span>
<span><span class="co">#&gt; 6: rs55664108 chr1 204587862 8.572e-25 0.7157  0.0912 0.0089 ENSG00000281406</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-3-conduct-colocalization-analysis-with-eqtl">Step 3: Conduct colocalization analysis with eQTL:<a class="anchor" aria-label="anchor" href="#step-3-conduct-colocalization-analysis-with-eqtl"></a>
</h2>
<p>Following three steps of colocalization analysis are encapsulated in
one function xQTLanalyze_coloc:</p>
<ol style="list-style-type: decimal">
<li>Retrieved all associations from EBI eQTL catalogue for a specified
gene.</li>
<li>Merge the data.frame of GWAS and eQTL by rsid.</li>
<li>Perform colocalization analysis using <code>coloc</code> and
<code>hyprcoloc</code> methods.</li>
</ol>
<p>For above 835 trait genes, a for loop can be used to get these genes’
outputs of colocalization analysis (this may take several minutes):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># All genes:</span></span>
<span><span class="va">genesAll</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLquery_gene.html">xQTLquery_gene</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">traitsAll</span><span class="op">$</span><span class="va">gencodeId</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setkey.html" class="external-link">setindex</a></span><span class="op">(</span><span class="va">gwasDF</span>, <span class="va">rsid</span><span class="op">)</span> <span class="co"># speed up merging</span></span>
<span><span class="co"># A data.table to save results:</span></span>
<span><span class="va">colocResultAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">genesAll</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># eQTL of the gene i</span></span>
<span>  <span class="va">eQTL_i</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLdownload_eqtlAllAsso.html">xQTLdownload_eqtlAllAsso</a></span><span class="op">(</span><span class="va">genesAll</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">$</span><span class="va">gencodeId</span>, geneType <span class="op">=</span> <span class="st">"gencodeId"</span>, tissueLabel<span class="op">=</span><span class="va">tissueSiteDetail</span>, withB37VariantId <span class="op">=</span> <span class="cn">FALSE</span>, data_source<span class="op">=</span><span class="st">"liLab"</span><span class="op">)</span></span>
<span>  <span class="va">gwasDF_i</span> <span class="op">&lt;-</span> <span class="va">gwasDF</span><span class="op">[</span><span class="va">rsid</span>  <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">eQTL_i</span><span class="op">$</span><span class="va">rsid</span>,<span class="op">]</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">eQTL_i</span><span class="op">)</span><span class="op">)</span><span class="op">{</span><span class="kw">next</span><span class="op">(</span><span class="op">)</span><span class="op">}</span>; <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gwasDF_i</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">{</span><span class="kw">next</span><span class="op">(</span><span class="op">)</span><span class="op">}</span></span>
<span>  <span class="va">eQTL_i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">eQTL_i</span>, <span class="va">gwasDF_i</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span><span class="op">)</span><span class="op">]</span>, by <span class="op">=</span><span class="st">"rsid"</span><span class="op">)</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue</span>, <span class="va">maf</span>, <span class="va">beta</span>, <span class="va">se</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="co"># using both method coloc and hyprcoloc to perform colocalization analysis</span></span>
<span>  <span class="va">colocResult_i</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_coloc_diy.html">xQTLanalyze_coloc_diy</a></span><span class="op">(</span>gwasDF <span class="op">=</span> <span class="va">gwasDF_i</span>, qtlDF <span class="op">=</span> <span class="va">eQTL_i</span>, method<span class="op">=</span><span class="st">"Both"</span><span class="op">)</span></span>
<span>  <span class="co"># </span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">colocResult_i</span><span class="op">)</span><span class="op">)</span><span class="op">{</span> </span>
<span>      <span class="va">colocResult_i</span> <span class="op">&lt;-</span> <span class="va">colocResult_i</span><span class="op">$</span><span class="va">coloc_Out_summary</span></span>
<span>      <span class="va">colocResult_i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">genesAll</span><span class="op">[</span><span class="va">i</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"geneSymbol"</span>,<span class="st">"gencodeId"</span><span class="op">)</span><span class="op">]</span>, <span class="va">colocResult_i</span><span class="op">)</span></span>
<span>      <span class="va">colocResultAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">colocResultAll</span>, <span class="va">colocResult_i</span><span class="op">)</span><span class="op">}</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">" == Id: "</span>,<span class="va">i</span>,<span class="st">"/"</span>,<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">genesAll</span><span class="op">)</span>,<span class="st">" == Gene:"</span>,<span class="va">genesAll</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">$</span><span class="va">gencodeId</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In this case, we invoke two methods to conduct colocalization
analysis, (1). <code>coloc</code> that estimates the posterior support
of variants for each hypothesis: H0,H1,H2,H3,H4:</p>
<ul>
<li>H0 : neither trait has a genetic association in the region</li>
<li>H1: only trait 1 has a genetic association in the region</li>
<li>H2: only trait 2 has a genetic association in the region</li>
<li>H3: both traits are associated, but with different causal
variants</li>
<li>H4: both traits are associated and share a single causal
variant</li>
</ul>
<p>and (2). <code>hyprcoloc</code> that calculates the posterior
probability that all traits share a causal variant.</p>
<p>Output is a data.table object that combined all results of
<code>colocOut</code> of all genes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">colocResultAll</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;       geneSymbol          gencodeId nsnps    PP.H0.abf    PP.H1.abf</span></span>
<span><span class="co">#&gt; 1: RP11-534L20.5  ENSG00000261000.1  4472 2.179842e-12 4.815048e-09</span></span>
<span><span class="co">#&gt; 2:         IKBKE  ENSG00000263528.7  4505 5.794749e-05 1.280001e-01</span></span>
<span><span class="co">#&gt; 3:        PM20D1 ENSG00000162877.12  5007 3.101352e-13 6.851499e-10</span></span>
<span><span class="co">#&gt; 4:         RAB29 ENSG00000117280.12  4998 2.696451e-11 5.957032e-08</span></span>
<span><span class="co">#&gt; 5:        NUCKS1 ENSG00000069275.12  5002 2.005928e-07 4.431533e-04</span></span>
<span><span class="co">#&gt; 6:       BLACAT1  ENSG00000281406.1  5011 4.875521e-19 4.875783e-01</span></span>
<span><span class="co">#&gt;       PP.H2.abf  PP.H3.abf    PP.H4.abf candidate_snp  SNP.PP.H4 hypr_posterior</span></span>
<span><span class="co">#&gt; 1: 4.522413e-04 0.99895403 0.0005937237     rs6693843 0.19259384         0.0001</span></span>
<span><span class="co">#&gt; 2: 3.939882e-04 0.87027865 0.0012693057      rs823121 0.07040552         0.0000</span></span>
<span><span class="co">#&gt; 3: 8.561368e-05 0.18832593 0.8115884566     rs1772159 0.39272007         0.5931</span></span>
<span><span class="co">#&gt; 4: 1.329558e-05 0.02840117 0.9715854781      rs708723 0.36167886         0.9428</span></span>
<span><span class="co">#&gt; 5: 1.253436e-05 0.02671831 0.9728258014      rs823154 0.42379078         0.9364</span></span>
<span><span class="co">#&gt; 6: 4.805752e-19 0.48056911 0.0318526046      rs930947 0.12798834         0.0039</span></span>
<span><span class="co">#&gt;    hypr_regional_prob hypr_candidate_snp hypr_posterior_explainedBySnp</span></span>
<span><span class="co">#&gt; 1:             0.2374          rs6693843                        0.2379</span></span>
<span><span class="co">#&gt; 2:             0.0057           rs823121                        0.0661</span></span>
<span><span class="co">#&gt; 3:             0.9995          rs1772159                        0.3313</span></span>
<span><span class="co">#&gt; 4:             1.0000           rs708723                        0.4030</span></span>
<span><span class="co">#&gt; 5:             0.9993           rs823154                        0.4423</span></span>
<span><span class="co">#&gt; 6:             0.0353          rs4245739                        0.1158</span></span></code></pre>
<p>We can see that the two methodologies’ outcomes are quite consistent.
With a threshold of 0.75 for <code>coloc</code> and 0.5 for
<code>hyprcoloc</code>, a total of 26 colocalized genes are shared.</p>
<p>To save time and go through this case as soon as possible, you can
get the above result directly with:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colocResultAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"http://bioinfo.szbl.ac.cn/xQTL_biolinks/xqtl_data/colocResultAll.txt"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-4-visualization-of-the-results">Step 4: Visualization of the results:<a class="anchor" aria-label="anchor" href="#step-4-visualization-of-the-results"></a>
</h2>
<p>We considered colocalization tests with a posterior probability of
hypothesis 4 (PPH4.ABF) &gt; 0.75 &amp; hypr_posterior &gt; 0.5 as
having strong or moderate evidence for colocalization.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colocResultsig</span> <span class="op">&lt;-</span> <span class="va">colocResultAll</span><span class="op">[</span><span class="va">PP.H4.abf</span><span class="op">&gt;</span><span class="fl">0.75</span> <span class="op">&amp;</span> <span class="va">hypr_posterior</span><span class="op">&gt;</span><span class="fl">0.5</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">PP.H4.abf</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>There are 26 trait genes that are associated and share a single
causal variant:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">colocResultsig</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;       geneSymbol          gencodeId nsnps    PP.H0.abf    PP.H1.abf</span></span>
<span><span class="co">#&gt; 1:          MMP7  ENSG00000137673.8  7108 4.037711e-17 8.285257e-05</span></span>
<span><span class="co">#&gt; 2: CTC-435M10.12  ENSG00000277744.1  4676 2.078933e-20 2.421163e-03</span></span>
<span><span class="co">#&gt; 3:      PPP1R14A ENSG00000167641.10  5034 5.328594e-34 2.988721e-07</span></span>
<span><span class="co">#&gt; 4:        GEMIN4 ENSG00000179409.10  6169 9.325967e-19 1.409617e-03</span></span>
<span><span class="co">#&gt; 5:          GGCX ENSG00000115486.11  5389 9.622300e-27 3.521949e-09</span></span>
<span><span class="co">#&gt; 6:          TBX1 ENSG00000184058.12  6044 1.851045e-09 1.456683e-07</span></span>
<span><span class="co">#&gt;       PP.H2.abf    PP.H3.abf PP.H4.abf candidate_snp SNP.PP.H4 hypr_posterior</span></span>
<span><span class="co">#&gt; 1: 6.360696e-16 0.0003055833 0.9996116    rs11568818 1.0000000         0.9993</span></span>
<span><span class="co">#&gt; 2: 1.588664e-20 0.0008534614 0.9967254     rs2079811 0.9999764         0.9923</span></span>
<span><span class="co">#&gt; 3: 8.166955e-30 0.0035842959 0.9964154     rs4802297 0.2683674         0.9911</span></span>
<span><span class="co">#&gt; 4: 2.672611e-18 0.0030440968 0.9955463     rs2740358 0.9493232         0.9869</span></span>
<span><span class="co">#&gt; 5: 2.411952e-20 0.0078360511 0.9921639    rs10175792 0.3177452         0.9830</span></span>
<span><span class="co">#&gt; 6: 2.233151e-04 0.0165906223 0.9831859     rs2238775 0.2005480         0.9778</span></span>
<span><span class="co">#&gt;    hypr_regional_prob hypr_candidate_snp hypr_posterior_explainedBySnp</span></span>
<span><span class="co">#&gt; 1:             0.9998         rs11568818                        1.0000</span></span>
<span><span class="co">#&gt; 2:             0.9943          rs2079811                        1.0000</span></span>
<span><span class="co">#&gt; 3:             1.0000          rs4802297                        0.2738</span></span>
<span><span class="co">#&gt; 4:             0.9960          rs2740358                        0.8927</span></span>
<span><span class="co">#&gt; 5:             1.0000         rs10175792                        0.3449</span></span>
<span><span class="co">#&gt; 6:             0.9995          rs2238775                        0.3230</span></span></code></pre>
<p>All these genes’ details can be fetched with
<code>xQTLquery_gene</code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLquery_gene.html">xQTLquery_gene</a></span><span class="op">(</span><span class="va">colocResultsig</span><span class="op">$</span><span class="va">gencodeId</span><span class="op">)</span></span></code></pre></div>
<p>Add the value of PPH4 for each gene, and remove non-protein-coding
genes:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">colocResultsig</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gencodeId"</span>, <span class="st">"PP.H4.abf"</span>, <span class="st">"candidate_snp"</span>, <span class="st">"SNP.PP.H4"</span>, <span class="st">"hypr_posterior"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                  <span class="va">outGenes</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"geneSymbol"</span>, <span class="st">"gencodeId"</span>, <span class="st">"entrezGeneId"</span>, <span class="st">"geneType"</span><span class="op">)</span><span class="op">]</span>, by<span class="op">=</span><span class="st">"gencodeId"</span>, sort<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">outGenes</span> <span class="op">&lt;-</span> <span class="va">outGenes</span><span class="op">[</span><span class="va">geneType</span> <span class="op">==</span><span class="st">"protein coding"</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outGenes</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;              gencodeId PP.H4.abf candidate_snp  SNP.PP.H4 hypr_posterior</span></span>
<span><span class="co">#&gt;  1:  ENSG00000137673.8 0.9996116    rs11568818 0.99999996         0.9993</span></span>
<span><span class="co">#&gt;  2: ENSG00000167641.10 0.9964154     rs4802297 0.26836743         0.9911</span></span>
<span><span class="co">#&gt;  3: ENSG00000179409.10 0.9955463     rs2740358 0.94932317         0.9869</span></span>
<span><span class="co">#&gt;  4: ENSG00000115486.11 0.9921639    rs10175792 0.31774517         0.9830</span></span>
<span><span class="co">#&gt;  5: ENSG00000184058.12 0.9831859     rs2238775 0.20054800         0.9778</span></span>
<span><span class="co">#&gt;  6: ENSG00000069275.12 0.9728258      rs823154 0.42379078         0.9364</span></span>
<span><span class="co">#&gt;  7: ENSG00000117280.12 0.9715855      rs708723 0.36167886         0.9428</span></span>
<span><span class="co">#&gt;  8: ENSG00000099331.13 0.9713063    rs11666569 0.25871794         0.9293</span></span>
<span><span class="co">#&gt;  9: ENSG00000184012.11 0.9660070     rs6517673 0.38159830         0.9127</span></span>
<span><span class="co">#&gt; 10:  ENSG00000172613.7 0.9576434    rs35826789 0.92558434         0.9038</span></span>
<span><span class="co">#&gt; 11: ENSG00000155749.12 0.9561720     rs7560328 0.22497345         0.8966</span></span>
<span><span class="co">#&gt; 12: ENSG00000204536.13 0.9366182     rs2517985 0.13110756         0.7333</span></span>
<span><span class="co">#&gt; 13:  ENSG00000180535.3 0.9361589     rs6465657 0.17430973         0.8643</span></span>
<span><span class="co">#&gt; 14: ENSG00000003400.14 0.9105858    rs12620010 0.37295865         0.7922</span></span>
<span><span class="co">#&gt; 15: ENSG00000101751.10 0.9084474     rs3730783 0.09578602         0.8321</span></span>
<span><span class="co">#&gt; 16: ENSG00000152601.17 0.8906783    rs73011743 0.03316094         0.7482</span></span>
<span><span class="co">#&gt; 17: ENSG00000136819.15 0.8812364    rs13300897 0.15189809         0.7313</span></span>
<span><span class="co">#&gt; 18: ENSG00000115648.13 0.8654407    rs72620824 0.54349674         0.7378</span></span>
<span><span class="co">#&gt; 19: ENSG00000198625.12 0.8209748     rs2169137 0.14030719         0.6086</span></span>
<span><span class="co">#&gt; 20: ENSG00000162877.12 0.8115885     rs1772159 0.39272007         0.5931</span></span>
<span><span class="co">#&gt; 21: ENSG00000065060.16 0.7949202   rs112283401 0.03978575         0.6114</span></span>
<span><span class="co">#&gt; 22: ENSG00000080546.13 0.7811190     rs6935782 0.08037725         0.6257</span></span>
<span><span class="co">#&gt; 23: ENSG00000118961.14 0.7700294     rs9306895 0.54356363         0.7300</span></span>
<span><span class="co">#&gt;              gencodeId PP.H4.abf candidate_snp  SNP.PP.H4 hypr_posterior</span></span>
<span><span class="co">#&gt;     geneSymbol entrezGeneId       geneType</span></span>
<span><span class="co">#&gt;  1:       MMP7         4316 protein coding</span></span>
<span><span class="co">#&gt;  2:   PPP1R14A        94274 protein coding</span></span>
<span><span class="co">#&gt;  3:     GEMIN4        50628 protein coding</span></span>
<span><span class="co">#&gt;  4:       GGCX         2677 protein coding</span></span>
<span><span class="co">#&gt;  5:       TBX1         6899 protein coding</span></span>
<span><span class="co">#&gt;  6:     NUCKS1        64710 protein coding</span></span>
<span><span class="co">#&gt;  7:      RAB29         8934 protein coding</span></span>
<span><span class="co">#&gt;  8:      MYO9B         4650 protein coding</span></span>
<span><span class="co">#&gt;  9:    TMPRSS2         7113 protein coding</span></span>
<span><span class="co">#&gt; 10:      RAD9A         5883 protein coding</span></span>
<span><span class="co">#&gt; 11:   ALS2CR12       130540 protein coding</span></span>
<span><span class="co">#&gt; 12:     CCHCR1        54535 protein coding</span></span>
<span><span class="co">#&gt; 13:    BHLHA15       168620 protein coding</span></span>
<span><span class="co">#&gt; 14:     CASP10          843 protein coding</span></span>
<span><span class="co">#&gt; 15:       POLI        11201 protein coding</span></span>
<span><span class="co">#&gt; 16:      MBNL1         4154 protein coding</span></span>
<span><span class="co">#&gt; 17:    C9orf78        51759 protein coding</span></span>
<span><span class="co">#&gt; 18:       MLPH        79083 protein coding</span></span>
<span><span class="co">#&gt; 19:       MDM4         4194 protein coding</span></span>
<span><span class="co">#&gt; 20:     PM20D1       148811 protein coding</span></span>
<span><span class="co">#&gt; 21:   UHRF1BP1        54887 protein coding</span></span>
<span><span class="co">#&gt; 22:      SESN1        27244 protein coding</span></span>
<span><span class="co">#&gt; 23:       LDAH        60526 protein coding</span></span>
<span><span class="co">#&gt;     geneSymbol entrezGeneId       geneType</span></span></code></pre>
<p>Ridgeline plot can be used to compare the expressions among these
genes:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_genesExp.html">xQTLvisual_genesExp</a></span><span class="op">(</span><span class="va">outGenes</span><span class="op">$</span><span class="va">geneSymbol</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_geneExpInOneTissue.png" width="50%" style="display: block; margin: auto;"></p>
<p>Trait gene <strong><code>MMP7</code></strong> that with the highest
<code>PPH4.ABF=0.9978</code> &amp; <code>hypr_posterior=0.9993</code>
encodes a member of the peptidase M10 family of matrix
metalloproteinases, which is involved in the breakdown of extracellular
matrix in normal physiological processes, such as embryonic development,
reproduction, and tissue remodeling, as well as in disease processes,
such as arthritis and metastasis. Prostate cancer can be promoted via
MMP7-induced epithelial-to-mesenchymal transition by Interleukin-17.
Resent literature has shown that serum MMP7 levels could guide
metastatic therapy for prostate cancer.</p>
<p>Expression of MMP7 in multiple tissues can be plotted with
<code>xQTLvisual_geneExpTissues</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneExpTissues</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLvisual_geneExpTissues.html">xQTLvisual_geneExpTissues</a></span><span class="op">(</span><span class="st">"MMP7"</span>, log10y <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_exp_multissues.png" width="90%" style="display: block; margin: auto;"></p>
<p>The number and significance of eQTLs in distinguished tissues are
capable of showing a tissue-specific effect or a ubiquitous effect. The
function <code>xQTLvisual_eqtl</code> can be used to indicate whether
the gene is widely regulation in various tissues.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_eqtl.html">xQTLvisual_eqtl</a></span><span class="op">(</span><span class="st">"MMP7"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_eqtlTissues.png" width="90%" style="display: block; margin: auto;"></p>
<p>Besides, we provide functions <code>xQTLvisual_locusCompare</code>
and <code>xQTLvisual_locusZoom</code> to visualize the colocalization
between the GWAS and the eQTL dataset for a specified gene, we take the
gene <code>MMP7</code> as an example:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download all eQTL associations of gene  MMP7 in prostate:</span></span>
<span><span class="va">eqtlAsso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLdownload_eqtlAllAsso.html">xQTLdownload_eqtlAllAsso</a></span><span class="op">(</span>gene<span class="op">=</span><span class="st">"MMP7"</span>,tissueLabel <span class="op">=</span> <span class="va">tissueSiteDetail</span><span class="op">)</span></span>
<span><span class="va">eqtlAsso</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/as.data.table.html" class="external-link">as.data.table</a></span><span class="op">(</span><span class="va">eqtlAsso</span><span class="op">)</span></span>
<span><span class="co"># Merge the variants of GWAS and eQTL dataset by rsid:</span></span>
<span><span class="va">gwasEqtldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AF"</span><span class="op">)</span><span class="op">]</span>, <span class="va">eqtlAsso</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span>rsid<span class="op">=</span><span class="va">snpId</span>, <span class="va">pValue</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                      by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span><span class="op">)</span>, suffixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gwas"</span>,<span class="st">".eqtl"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Five retained fields are required:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwasEqtldata</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;             rsid chrom  position pValue.gwas    beta     se pValue.eqtl</span></span>
<span><span class="co">#&gt;    1:  rs1004352    11 101777151   0.1901000 -0.0108 0.0083    0.175397</span></span>
<span><span class="co">#&gt;    2: rs10047467    11 101761011   0.4854000 -0.0096 0.0138    0.870120</span></span>
<span><span class="co">#&gt;    3: rs10047477    11 101696584   0.0791000  0.0147 0.0084    0.570213</span></span>
<span><span class="co">#&gt;    4: rs10047501    11 101696582   0.0755500 -0.0148 0.0083    0.570213</span></span>
<span><span class="co">#&gt;    5:  rs1006330    11 101626601   0.0564300  0.0169 0.0089    0.715533</span></span>
<span><span class="co">#&gt;   ---                                                                  </span></span>
<span><span class="co">#&gt; 7530:   rs996999    11 102667063   0.7268000 -0.0036 0.0102    0.309526</span></span>
<span><span class="co">#&gt; 7531:   rs998379    11 101587510   0.8520000 -0.0024 0.0126    0.156798</span></span>
<span><span class="co">#&gt; 7532:  rs9988826    11 101673906   0.7874000  0.0027 0.0101    0.117728</span></span>
<span><span class="co">#&gt; 7533:  rs9988861    11 101438758   0.0006782 -0.0306 0.0090    0.404080</span></span>
<span><span class="co">#&gt; 7534:  rs9988919    11 102876923   0.2217000  0.0331 0.0271    0.020295</span></span></code></pre>
<p>Visualization of p-value distribution and comparison of the signals
of GWAS and eQTL:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusCompare.html">xQTLvisual_locusCompare</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">pValue.eqtl</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                        <span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">pValue.gwas</span><span class="op">)</span><span class="op">]</span>, legend_position <span class="op">=</span> <span class="st">"bottomright"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_locuscompare.png" width="50%" style="display: block; margin: auto;"></p>
<p>Locuszoom plot of GWAS signals:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusZoom.html">xQTLvisual_locusZoom</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue.gwas</span><span class="op">)</span><span class="op">]</span>, legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_locuszoom_gwas.png" width="50%" style="display: block; margin: auto;"></p>
<p>Locuszoom plot of eQTL signals:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusZoom.html">xQTLvisual_locusZoom</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue.eqtl</span><span class="op">)</span><span class="op">]</span>, legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_locuszoom_eqtl.png" width="50%" style="display: block; margin: auto;"></p>
<p>We can also combine locuscompare and locuszoom plot using function
<code>xQTLvisual_locusCombine</code>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusCombine.html">xQTLvisual_locusCombine</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>,<span class="st">"chrom"</span>, <span class="st">"position"</span>, <span class="st">"pValue.gwas"</span>, <span class="st">"pValue.eqtl"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_combined.png" width="90%" style="display: block; margin: auto;"></p>
<p>From the above figures, we can see that the SNP
<code>rs11568818</code> is potential causal variant, and we can use a
violin plot to show the normalized effect size of it:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_eqtlExp.html">xQTLvisual_eqtlExp</a></span><span class="op">(</span><span class="st">"rs11568818"</span>, <span class="st">"MMP7"</span>, tissueSiteDetail <span class="op">=</span> <span class="va">tissueSiteDetail</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_eQTL_exp.png" width="40%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="extension-integrative-analysis-with-external-packages-or-tools">Extension: Integrative analysis with external packages or tools<a class="anchor" aria-label="anchor" href="#extension-integrative-analysis-with-external-packages-or-tools"></a>
</h2>
<p>To gain insight into the function of trait genes with higher PPH4,
and explore the potential regulatory mechanism of the prostate cancer,
we can conduct exploratory analysis, like co-expression analysis, and
gene ontology enrichment analyses.</p>
<p>First we download expression profiles of the genes with higher value
of PPH4 (&gt;0.75) &amp; hypr_posterior (&gt;0.5) in prostate.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLdownload_exp.html">xQTLdownload_exp</a></span><span class="op">(</span><span class="va">outGenes</span><span class="op">$</span><span class="va">gencodeId</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span>, toSummarizedExperiment <span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Pearson coefficient can be calculated with the expression matrix for
each gene:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corDT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">expMat</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">:</span><span class="op">-</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">corDT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">outGenes</span><span class="op">$</span><span class="va">geneSymbol</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">corDT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">outGenes</span><span class="op">$</span><span class="va">geneSymbol</span></span></code></pre></div>
<p>R package <code>corrplot</code> is used to display this correlation
matrix:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/taiyun/corrplot" class="external-link">corrplot</a></span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html" class="external-link">corrplot</a></span><span class="op">(</span><span class="va">corDT</span>, method<span class="op">=</span><span class="st">"color"</span>,  </span>
<span>         type<span class="op">=</span><span class="st">"upper"</span>,    </span>
<span>         order <span class="op">=</span> <span class="st">"hclust"</span>,</span>
<span>         addCoef.col <span class="op">=</span> <span class="st">"#ff0099"</span>, </span>
<span>         number.cex <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> </span></code></pre></div>
<p><img src="images/prostate/prostate_cor.png" width="90%" style="display: block; margin: auto;"></p>
<p>R package <code>clusterProfiler</code> is used for gene functional
annotation:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">ego</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">outGenes</span><span class="op">$</span><span class="va">entrezGeneId</span><span class="op">)</span>,</span>
<span>                OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>                ont<span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                pAdjustMethod <span class="op">=</span><span class="st">"none"</span>,</span>
<span>                readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">ego</span>, showCategory<span class="op">=</span><span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_Go.png" width="80%" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ruofan Ding, Xudong Zou, Yangmei Qin, Gao Wang, Lei Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
