<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Case study: colocalization analysis in prostate cancer • xQTLbiolinks</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Case study: colocalization analysis in prostate cancer">
<meta property="og:description" content="xQTLbiolinks">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xQTLbiolinks</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/visualization.html">Visualization</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dingruofan/xQTLbiolinks/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Case study: colocalization analysis in prostate
cancer</h1>
            
            <h4 data-toc-skip class="date">2022-06-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dingruofan/xQTLbiolinks/blob/HEAD/vignettes/Colocalization_analysis_with_xQTLbiolinks.Rmd" class="external-link"><code>vignettes/Colocalization_analysis_with_xQTLbiolinks.Rmd</code></a></small>
      <div class="hidden name"><code>Colocalization_analysis_with_xQTLbiolinks.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="before-we-start-">Before we start.<a class="anchor" aria-label="anchor" href="#before-we-start-"></a>
</h2>
<p>All we need to prepare include three parts:</p>
<ol style="list-style-type: decimal">
<li>GWAS summary statistics dataset.</li>
<li>Name of tissue of interest.</li>
<li>Following three installed packages.</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dingruofan/xQTLbiolinks" class="external-link">xQTLbiolinks</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span></code></pre></div>
<p>Prostate cancer is one of the most common cancers in men. Prostate
cancer pathogenesis involves both heritable and environmental factors.
The molecular events involved in the development or progression of
prostate cancer are still unclear. In this example, we aim to figure out
the causal variants and genes assocaited with prostate cancer, and to
uncover potential molecular mechanisms of regulation.</p>
<p>For data preparation, we download summary statistics dataset of a
GWAS study (<a href="http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST006001-GCST007000/GCST006085/harmonised/29892016-GCST006085-EFO_0001663-build37.f.tsv.gz" class="external-link">GCST006085</a>)
of prostate cancer from GWAS category and load the dataset in R with
<code>data.table</code> package. Correspondingly, we chose tissue
<code>Prostate</code> for study. We retain the variants with dbSNP id
(start with <code>rs</code>), and a data.table object named
<code>gwasDF</code> of 13,498,990 (rows) x 7 (cols) is loaded.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwasDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"29892016-GCST006085-EFO_0001663-build37.f.tsv.gz"</span><span class="op">)</span></span>
<span><span class="co"># extract columns.</span></span>
<span><span class="va">gwasDF</span><span class="op">&lt;-</span> <span class="va">gwasDF</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">variant_id</span>, <span class="st">"^rs"</span><span class="op">)</span>,<span class="fu">.</span><span class="op">(</span>rsid<span class="op">=</span><span class="va">variant_id</span>, chrom<span class="op">=</span><span class="va">chromosome</span>, position<span class="op">=</span> <span class="va">base_pair_location</span>, pValue<span class="op">=</span><span class="va">p_value</span>, AF<span class="op">=</span><span class="va">effect_allele_frequency</span>, <span class="va">beta</span>, se<span class="op">=</span> <span class="va">standard_error</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># tissue:</span></span>
<span><span class="va">tissueSiteDetail</span><span class="op">=</span><span class="st">"Prostate"</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;           rsid chrom position  pValue     AF    beta     se</span></span>
<span><span class="co">#&gt; 1: rs145072688     1    10352 0.46280 0.5753 -0.0094 0.0128</span></span>
<span><span class="co">#&gt; 2: rs376342519     1    10616 0.41070 0.0066  0.0572 0.0696</span></span>
<span><span class="co">#&gt; 3: rs201725126     1    13116 0.84470 0.8255 -0.0035 0.0178</span></span>
<span><span class="co">#&gt; 4: rs200579949     1    13118 0.84470 0.8255 -0.0035 0.0178</span></span>
<span><span class="co">#&gt; 5:  rs75454623     1    14930 0.61880 0.4772 -0.0065 0.0131</span></span>
<span><span class="co">#&gt; 6: rs199856693     1    14933 0.02721 0.0407 -0.0794 0.0360</span></span></code></pre>
<p>Now we can perform quality control for this GWAS dataset by
estimating the genomic inflation factor, validating QQ-plot and PZ-plot
using xQTLbiolinks.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_qqPlot.html">xQTLvisual_qqPlot</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">pValue</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/xQTLvisual_PZPlot.html">xQTLvisual_PZPlot</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">pValue</span>, <span class="va">beta</span>, <span class="va">se</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/qq_pz_plot.png" width="50%" style="display: block; margin: auto;">
Left is the QQ plot. The x-axis of the QQ-plot represents the expected
p-values under the null hypothesis, while the y-axis represents the
observed p-values from the GWAS summary statistics data. Right is the PZ
plot, he observed p-values on the y-axis and the corresponding p-values
derived from z-scores that estimated by beta and se on the x-axis.</p>
</div>
<div class="section level2">
<h2 id="step-1-filter-sentinel-snps-">Step 1: Filter sentinel snps.<a class="anchor" aria-label="anchor" href="#step-1-filter-sentinel-snps-"></a>
</h2>
<p>Sentinel SNP is the most prominent signal within a given genome
range, and is usually in high LD with causal variants. By default,
xQTLbiolinks detect sentinel snps that with the p-value &lt; 5e-8 and
SNP-to-SNP distance &gt; 10e6 bp. <strong>Note</strong>: For in this
example, due to the inconsistent genome version between the GWAS dataset
(GRCh37) and eQTL associations (GRCh38) from eQTL category, conversion
of genome version is required, and can be conducted using
<code>xQTLanalyze_getSentinelSnp</code> with
<code>genomeVersion="grch37"</code> and <code>grch37To38=TRUE</code>
(package <code>rtracklayer</code>is required):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sentinelSnpDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_getSentinelSnp.html">xQTLanalyze_getSentinelSnp</a></span><span class="op">(</span><span class="va">gwasDF</span>, centerRange<span class="op">=</span><span class="fl">1e6</span>,</span>
<span>                                            genomeVersion<span class="op">=</span><span class="st">"grch37"</span>, grch37To38<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>A total of 94 sentinel SNPs are detected.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sentinelSnpDF</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;           rsid  chr  position    pValue    maf    beta     se</span></span>
<span><span class="co">#&gt; 1:  rs55664108 chr1 204587862 8.572e-25 0.7157  0.0912 0.0089</span></span>
<span><span class="co">#&gt; 2:  rs35296356 chr1 150601662 3.355e-14 0.3384 -0.0690 0.0091</span></span>
<span><span class="co">#&gt; 3:  rs34579442 chr1 153927424 4.478e-14 0.6637 -0.0655 0.0087</span></span>
<span><span class="co">#&gt; 4: rs146564277 chr1 155051619 1.577e-12 0.0301 -0.1850 0.0262</span></span>
<span><span class="co">#&gt; 5:  rs34848415 chr1 205762038 2.893e-09 0.5231 -0.0501 0.0084</span></span>
<span><span class="co">#&gt; 6:  rs56391074 chr1  87745032 1.659e-08 0.6298 -0.0466 0.0082</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-2-identify-trait-genes-for-each-sentinel-snps-">Step 2: Identify trait genes for each sentinel SNPs.<a class="anchor" aria-label="anchor" href="#step-2-identify-trait-genes-for-each-sentinel-snps-"></a>
</h2>
<p>Trait genes are genes that located in the range of 1Mb (default, can
be changed with parameter <code>detectRange</code>) of sentinel SNPs. In
order to reduce the number of trait genes and thus reduce the running
time, we take the overlap of eGenes and trait genes as the final output
of the function <code>xQTLanalyze_getTraits</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traitsAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_getTraits.html">xQTLanalyze_getTraits</a></span><span class="op">(</span><span class="va">sentinelSnpDF</span>, detectRange<span class="op">=</span><span class="fl">1e6</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span></span></code></pre></div>
<p>Totally, 898 associations between 835 traits genes and 92 sentinel
SNPs are detected</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">traitsAll</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;    chromosome geneStart   geneEnd geneStrand geneSymbol          gencodeId</span></span>
<span><span class="co">#&gt; 1:       chr1 205086142 205122015          -      RBBP5 ENSG00000117222.13</span></span>
<span><span class="co">#&gt; 2:       chr1 205142505 205211566          -      DSTYK ENSG00000133059.16</span></span>
<span><span class="co">#&gt; 3:       chr1 205336065 205357090          -    KLHDC8A ENSG00000162873.14</span></span>
<span><span class="co">#&gt; 4:       chr1 204828651 205022822          +      NFASC ENSG00000163531.15</span></span>
<span><span class="co">#&gt; 5:       chr1 204190341 204196486          -      KISS1  ENSG00000170498.8</span></span>
<span><span class="co">#&gt; 6:       chr1 204198160 204214092          -     GOLT1A  ENSG00000174567.7</span></span>
<span><span class="co">#&gt;          rsid  position    pValue    maf</span></span>
<span><span class="co">#&gt; 1: rs55664108 204587862 8.572e-25 0.7157</span></span>
<span><span class="co">#&gt; 2: rs55664108 204587862 8.572e-25 0.7157</span></span>
<span><span class="co">#&gt; 3: rs55664108 204587862 8.572e-25 0.7157</span></span>
<span><span class="co">#&gt; 4: rs55664108 204587862 8.572e-25 0.7157</span></span>
<span><span class="co">#&gt; 5: rs55664108 204587862 8.572e-25 0.7157</span></span>
<span><span class="co">#&gt; 6: rs55664108 204587862 8.572e-25 0.7157</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-3-conduct-colocalization-analysis-">Step 3: Conduct colocalization analysis.<a class="anchor" aria-label="anchor" href="#step-3-conduct-colocalization-analysis-"></a>
</h2>
<p>Following three steps of colocalization analysis are encapsulated in
one function xQTLanalyze_coloc:</p>
<ol style="list-style-type: decimal">
<li>Retrieved all associations from EBI eQTL catalogue for a specified
gene.</li>
<li>Merge the data.frame of GWAS and eQTL by rsid.</li>
<li>Perform colocalization analysis using <code>coloc</code> and
<code>hyprcoloc</code> methods.</li>
</ol>
<p>For above 835 trait genes, a for loop can be used to get these genes’
outputs of colocalization analysis (this may take several hours):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># All genes:</span></span>
<span><span class="va">genesAll</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">traitsAll</span><span class="op">$</span><span class="va">gencodeId</span><span class="op">)</span></span>
<span><span class="co"># A data.table to save results:</span></span>
<span><span class="va">colocResultAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genesAll</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">colocResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_coloc.html">xQTLanalyze_coloc</a></span><span class="op">(</span><span class="va">gwasDF</span>, </span>
<span>                                   genomeVersion <span class="op">=</span> <span class="st">"grch37"</span>, method<span class="op">=</span><span class="st">"Both"</span>,</span>
<span>                                   <span class="va">genesAll</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">colocResult</span><span class="op">)</span><span class="op">)</span><span class="op">{</span> <span class="va">colocResultAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">colocResultAll</span>, <span class="va">colocResult</span><span class="op">$</span><span class="va">colocOut</span><span class="op">)</span><span class="op">}</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"== %Y-%b-%d %H:%M:%S "</span><span class="op">)</span>,<span class="st">" == Id:"</span>,<span class="va">i</span>,<span class="st">"/"</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genesAll</span><span class="op">)</span>,<span class="st">" == Gene:"</span>,<span class="va">genesAll</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In this case, we invoke two methods to conduct colocalization
analysis, (1). <code>coloc</code> that estimates the posterior support
of variants for each hypothesis: H0,H1,H2,H3,H4:</p>
<ul>
<li>H0 : neither trait has a genetic association in the region</li>
<li>H1: only trait 1 has a genetic association in the region</li>
<li>H2: only trait 2 has a genetic association in the region</li>
<li>H3: both traits are associated, but with different causal
variants</li>
<li>H4: both traits are associated and share a single causal
variant</li>
</ul>
<p>and (2). <code>hyprcoloc</code> that calculates the posterior
probability that all traits share a causal variant.</p>
<p>Output is a data.table object that combined all results of
<code>colocOut</code> of all genes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">colocResultAll</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;             traitGene nsnps    PP.H0.abf    PP.H1.abf    PP.H2.abf  PP.H3.abf</span></span>
<span><span class="co">#&gt; 1: ENSG00000001167.14  6205 2.072318e-15 0.2592617382 5.785196e-15 0.72375223</span></span>
<span><span class="co">#&gt; 2: ENSG00000002919.14  4883 2.252628e-42 0.0004579085 4.916915e-39 0.99949789</span></span>
<span><span class="co">#&gt; 3: ENSG00000003393.14  3901 4.134900e-03 0.5826462177 2.588394e-03 0.36468313</span></span>
<span><span class="co">#&gt; 4: ENSG00000003400.14  4549 7.800693e-05 0.0111580704 6.692117e-04 0.09483042</span></span>
<span><span class="co">#&gt; 5: ENSG00000005020.12  5271 1.551832e-32 0.4907383003 1.308029e-32 0.41354432</span></span>
<span><span class="co">#&gt; 6: ENSG00000005379.15  4413 9.273629e-03 0.4357011132 1.079004e-02 0.50690919</span></span>
<span><span class="co">#&gt;       PP.H4.abf candidate_snp SNP.PP.H4 hypr_posterior hypr_regional_prob</span></span>
<span><span class="co">#&gt; 1: 1.698603e-02     rs9369290 0.2487894         0.0000             0.0079</span></span>
<span><span class="co">#&gt; 2: 4.420283e-05   rs138467395 1.0000000         0.0000             0.0174</span></span>
<span><span class="co">#&gt; 3: 4.594735e-02    rs59308963 0.1405084         0.0003             0.0126</span></span>
<span><span class="co">#&gt; 4: 8.932643e-01    rs12620010 0.2315255         0.7922             0.9931</span></span>
<span><span class="co">#&gt; 5: 9.571738e-02    rs17156041 0.6731519         0.0005             0.0373</span></span>
<span><span class="co">#&gt; 6: 3.732603e-02     rs2189681 0.3030656         0.0002             0.0129</span></span>
<span><span class="co">#&gt;    hypr_candidate_snp hypr_posterior_explainedBySnp</span></span>
<span><span class="co">#&gt; 1:          rs9369290                        0.2752</span></span>
<span><span class="co">#&gt; 2:        rs138467395                        1.0000</span></span>
<span><span class="co">#&gt; 3:         rs59308963                        0.1434</span></span>
<span><span class="co">#&gt; 4:         rs12620010                        0.3882</span></span>
<span><span class="co">#&gt; 5:         rs17156041                        0.6855</span></span>
<span><span class="co">#&gt; 6:          rs2189681                        0.1829</span></span></code></pre>
<p>We can see that the two methodologies’ outcomes are quite consistent.
With a threshold of 0.75 for <code>coloc</code> and 0.5 for
<code>hyprcoloc</code>, a total of 24 colocalized genes are shared.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VennDiagram</span><span class="op">)</span></span>
<span><span class="cn">T</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VennDiagram/man/venn.diagram.html" class="external-link">venn.diagram</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>colocOut <span class="op">=</span> <span class="va">colocResultAll</span><span class="op">[</span><span class="va">PP.H4.abf</span><span class="op">&gt;=</span><span class="fl">0.75</span><span class="op">]</span><span class="op">$</span><span class="va">traitGene</span>, </span>
<span>                   hyprcolocOut <span class="op">=</span> <span class="va">colocResultAll</span><span class="op">[</span><span class="va">hypr_posterior</span><span class="op">&gt;=</span><span class="fl">0.5</span><span class="op">]</span><span class="op">$</span><span class="va">traitGene</span><span class="op">)</span>,</span>
<span>                   filename <span class="op">=</span> <span class="cn">NULL</span>,lwd<span class="op">=</span><span class="fl">1</span>,lty<span class="op">=</span><span class="fl">1</span>, width <span class="op">=</span> <span class="fl">8000</span>, height <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                   fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'#ff7a45'</span>,<span class="st">'#597ef7'</span><span class="op">)</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'red'</span>,<span class="st">'blue'</span><span class="op">)</span>, cat.col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'red'</span>,<span class="st">'blue'</span><span class="op">)</span>,</span>
<span>                   main.cex <span class="op">=</span> <span class="fl">0.45</span>, sub.cex<span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                   disable.logging <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/grid.draw.html" class="external-link">grid.draw</a></span><span class="op">(</span><span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/venn_coloc_hyprcoloc.png" width="30%" style="display: block; margin: auto;"></p>
<p>To save time and go through this case as soon as possible, you can
get the above result directly with:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colocResultAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"http://raw.githubusercontent.com/dingruofan/exampleData/master/colocResultAll.txt"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-4-visualization-of-the-results-">Step 4: Visualization of the results.<a class="anchor" aria-label="anchor" href="#step-4-visualization-of-the-results-"></a>
</h2>
<p>We considered colocalization tests with a posterior probability of
hypothesis 4 (PPH4.ABF) &gt; 0.75 &amp; hypr_posterior &gt; 0.5 as
having strong or moderate evidence for colocalization.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colocResultsig</span> <span class="op">&lt;-</span> <span class="va">colocResultAll</span><span class="op">[</span><span class="va">PP.H4.abf</span><span class="op">&gt;</span><span class="fl">0.75</span> <span class="op">&amp;</span> <span class="va">hypr_posterior</span><span class="op">&gt;</span><span class="fl">0.5</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">PP.H4.abf</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>There are 24 trait genes that are associated and share a single
causal variant:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">colocResultsig</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;             traitGene nsnps    PP.H0.abf    PP.H1.abf    PP.H2.abf   PP.H3.abf</span></span>
<span><span class="co">#&gt; 1:  ENSG00000137673.8  7108 3.722120e-16 7.637674e-04 1.159909e-15 0.001382244</span></span>
<span><span class="co">#&gt; 2: ENSG00000167641.10  5034 2.405778e-32 1.349362e-05 8.271165e-30 0.003642818</span></span>
<span><span class="co">#&gt; 3: ENSG00000184058.12  6045 1.551367e-07 1.220850e-05 1.078421e-04 0.007494261</span></span>
<span><span class="co">#&gt; 4: ENSG00000115486.11  5389 1.594413e-24 5.835862e-07 2.875803e-20 0.009535535</span></span>
<span><span class="co">#&gt; 5: ENSG00000179409.10  6154 4.387457e-18 6.631626e-03 6.233624e-18 0.008437170</span></span>
<span><span class="co">#&gt; 6:  ENSG00000277744.1  4616 1.043225e-19 1.214959e-02 6.298351e-20 0.006353677</span></span>
<span><span class="co">#&gt;    PP.H4.abf candidate_snp SNP.PP.H4 hypr_posterior hypr_regional_prob</span></span>
<span><span class="co">#&gt; 1: 0.9978540    rs11568818 0.9999999         0.9993             0.9998</span></span>
<span><span class="co">#&gt; 2: 0.9963437     rs7247241 0.2491879         0.9911             1.0000</span></span>
<span><span class="co">#&gt; 3: 0.9923855     rs2238775 0.3010829         0.9778             0.9995</span></span>
<span><span class="co">#&gt; 4: 0.9904639    rs10175792 0.2836228         0.9830             1.0000</span></span>
<span><span class="co">#&gt; 5: 0.9849312     rs2740358 0.9401973         0.9869             0.9960</span></span>
<span><span class="co">#&gt; 6: 0.9814967     rs2079811 0.9999034         0.9924             0.9943</span></span>
<span><span class="co">#&gt;    hypr_candidate_snp hypr_posterior_explainedBySnp</span></span>
<span><span class="co">#&gt; 1:         rs11568818                        1.0000</span></span>
<span><span class="co">#&gt; 2:          rs4802297                        0.2738</span></span>
<span><span class="co">#&gt; 3:          rs2238775                        0.3230</span></span>
<span><span class="co">#&gt; 4:         rs10175792                        0.3449</span></span>
<span><span class="co">#&gt; 5:          rs2740358                        0.8927</span></span>
<span><span class="co">#&gt; 6:          rs2079811                        1.0000</span></span></code></pre>
<p>All these genes’ details can be fetched with
<code>xQTLquery_gene</code>:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLquery_gene.html">xQTLquery_gene</a></span><span class="op">(</span><span class="va">colocResultsig</span><span class="op">$</span><span class="va">traitGene</span><span class="op">)</span></span></code></pre></div>
<p>Add the value of PPH4 for each gene, and remove non-protein-coding
genes:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">colocResultsig</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span>gencodeId<span class="op">=</span> <span class="va">traitGene</span>, <span class="va">PP.H4.abf</span>, <span class="va">candidate_snp</span>, <span class="va">SNP.PP.H4</span>, <span class="va">hypr_posterior</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                  <span class="va">outGenes</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">geneSymbol</span>, <span class="va">gencodeId</span>, <span class="va">entrezGeneId</span>, <span class="va">geneType</span><span class="op">)</span><span class="op">]</span>, by<span class="op">=</span><span class="st">"gencodeId"</span>, sort<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">outGenes</span> <span class="op">&lt;-</span> <span class="va">outGenes</span><span class="op">[</span><span class="va">geneType</span> <span class="op">==</span><span class="st">"protein coding"</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outGenes</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;              gencodeId PP.H4.abf candidate_snp  SNP.PP.H4 hypr_posterior</span></span>
<span><span class="co">#&gt;  1:  ENSG00000137673.8 0.9978540    rs11568818 0.99999989         0.9993</span></span>
<span><span class="co">#&gt;  2: ENSG00000167641.10 0.9963437     rs7247241 0.24918793         0.9911</span></span>
<span><span class="co">#&gt;  3: ENSG00000184058.12 0.9923855     rs2238775 0.30108289         0.9778</span></span>
<span><span class="co">#&gt;  4: ENSG00000115486.11 0.9904639    rs10175792 0.28362278         0.9830</span></span>
<span><span class="co">#&gt;  5: ENSG00000179409.10 0.9849312     rs2740358 0.94019727         0.9869</span></span>
<span><span class="co">#&gt;  6: ENSG00000117280.12 0.9701717      rs708723 0.31910829         0.9428</span></span>
<span><span class="co">#&gt;  7: ENSG00000184012.11 0.9652282     rs6517673 0.37410718         0.9127</span></span>
<span><span class="co">#&gt;  8: ENSG00000069275.12 0.9625282      rs823154 0.29904947         0.9364</span></span>
<span><span class="co">#&gt;  9: ENSG00000099331.13 0.9607259    rs11666569 0.24161844         0.9293</span></span>
<span><span class="co">#&gt; 10: ENSG00000155749.12 0.9538274     rs7560328 0.20927651         0.8966</span></span>
<span><span class="co">#&gt; 11: ENSG00000118961.14 0.9294589     rs9306895 0.48924035         0.7300</span></span>
<span><span class="co">#&gt; 12: ENSG00000101751.10 0.9271138     rs3730783 0.08842167         0.8321</span></span>
<span><span class="co">#&gt; 13:  ENSG00000172613.7 0.9099755    rs35826789 0.69464894         0.9038</span></span>
<span><span class="co">#&gt; 14:  ENSG00000180535.3 0.8941134     rs6465657 0.15229260         0.8643</span></span>
<span><span class="co">#&gt; 15: ENSG00000003400.14 0.8932643    rs12620010 0.23152548         0.7922</span></span>
<span><span class="co">#&gt; 16: ENSG00000115648.13 0.8889021    rs72620824 0.35924147         0.7378</span></span>
<span><span class="co">#&gt; 17: ENSG00000162877.12 0.8849219     rs1772159 0.28436834         0.5931</span></span>
<span><span class="co">#&gt; 18: ENSG00000204536.13 0.8733027     rs2517985 0.12887177         0.7333</span></span>
<span><span class="co">#&gt; 19: ENSG00000136819.15 0.8365468    rs13300897 0.12522776         0.7313</span></span>
<span><span class="co">#&gt; 20: ENSG00000065060.16 0.7937810   rs112283401 0.02533245         0.6114</span></span>
<span><span class="co">#&gt; 21: ENSG00000198625.12 0.7695907     rs2169137 0.12300108         0.6086</span></span>
<span><span class="co">#&gt;              gencodeId PP.H4.abf candidate_snp  SNP.PP.H4 hypr_posterior</span></span>
<span><span class="co">#&gt;     geneSymbol entrezGeneId       geneType</span></span>
<span><span class="co">#&gt;  1:       MMP7         4316 protein coding</span></span>
<span><span class="co">#&gt;  2:   PPP1R14A        94274 protein coding</span></span>
<span><span class="co">#&gt;  3:       TBX1         6899 protein coding</span></span>
<span><span class="co">#&gt;  4:       GGCX         2677 protein coding</span></span>
<span><span class="co">#&gt;  5:     GEMIN4        50628 protein coding</span></span>
<span><span class="co">#&gt;  6:      RAB29         8934 protein coding</span></span>
<span><span class="co">#&gt;  7:    TMPRSS2         7113 protein coding</span></span>
<span><span class="co">#&gt;  8:     NUCKS1        64710 protein coding</span></span>
<span><span class="co">#&gt;  9:      MYO9B         4650 protein coding</span></span>
<span><span class="co">#&gt; 10:   ALS2CR12       130540 protein coding</span></span>
<span><span class="co">#&gt; 11:       LDAH        60526 protein coding</span></span>
<span><span class="co">#&gt; 12:       POLI        11201 protein coding</span></span>
<span><span class="co">#&gt; 13:      RAD9A         5883 protein coding</span></span>
<span><span class="co">#&gt; 14:    BHLHA15       168620 protein coding</span></span>
<span><span class="co">#&gt; 15:     CASP10          843 protein coding</span></span>
<span><span class="co">#&gt; 16:       MLPH        79083 protein coding</span></span>
<span><span class="co">#&gt; 17:     PM20D1       148811 protein coding</span></span>
<span><span class="co">#&gt; 18:     CCHCR1        54535 protein coding</span></span>
<span><span class="co">#&gt; 19:    C9orf78        51759 protein coding</span></span>
<span><span class="co">#&gt; 20:   UHRF1BP1        54887 protein coding</span></span>
<span><span class="co">#&gt; 21:       MDM4         4194 protein coding</span></span>
<span><span class="co">#&gt;     geneSymbol entrezGeneId       geneType</span></span></code></pre>
<p>Ridgeline plot can be used to compare the expressions among these
genes:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_genesExp.html">xQTLvisual_genesExp</a></span><span class="op">(</span><span class="va">outGenes</span><span class="op">$</span><span class="va">geneSymbol</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_geneExpInOneTissue.png" width="50%" style="display: block; margin: auto;"></p>
<p>Trait gene <strong><code>MMP7</code></strong> that with the highest
<code>PPH4.ABF=0.9978</code> &amp; <code>hypr_posterior=0.9993</code>
encodes a member of the peptidase M10 family of matrix
metalloproteinases, which is involved in the breakdown of extracellular
matrix in normal physiological processes, such as embryonic development,
reproduction, and tissue remodeling, as well as in disease processes,
such as arthritis and metastasis <span class="citation">(<a href="#ref-RefSeq" role="doc-biblioref"><span>“RefSeq,”</span>
n.d.</a>)</span>. Prostate cancer can be promoted via MMP7-induced
epithelial-to-mesenchymal transition by Interleukin-17 <span class="citation">(<a href="#ref-zhang2017" role="doc-biblioref">Zhang et
al. 2017</a>)</span>. Resent literature has shown that serum MMP7 levels
could guide metastatic therapy for prostate cancer <span class="citation">(<a href="#ref-tregunna2020" role="doc-biblioref">Tregunna 2020</a>)</span>.</p>
<p>Expression of MMP7 in multiple tissues can be plotted with
<code>xQTLvisual_geneExpTissues</code>:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneExpTissues</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLvisual_geneExpTissues.html">xQTLvisual_geneExpTissues</a></span><span class="op">(</span><span class="st">"MMP7"</span>, log10y <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_exp_multissues.png" width="90%" style="display: block; margin: auto;"></p>
<p>The number and significance of eQTLs in distinguished tissues are
capable of showing a tissue-specific effect or a ubiquitous effect. The
function <code>xQTLvisual_eqtl</code> can be used to indicate whether
the gene is widely regulation in various tissues.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_eqtl.html">xQTLvisual_eqtl</a></span><span class="op">(</span><span class="st">"MMP7"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_eqtlTissues.png" width="90%" style="display: block; margin: auto;"></p>
<p>Besides, we provide functions <code>xQTLvisual_locusCompare</code>
and <code>xQTLvisual_locusZoom</code> to visualize the colocalization
between the GWAS and the eQTL dataset for a specified gene, we take the
gene <code>MMP7</code> as an example:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download all eQTL associations of gene  MMP7 in prostate:</span></span>
<span><span class="va">eqtlAsso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLdownload_eqtlAllAsso.html">xQTLdownload_eqtlAllAsso</a></span><span class="op">(</span>gene<span class="op">=</span><span class="st">"MMP7"</span>,tissueLabel <span class="op">=</span> <span class="va">tissueSiteDetail</span><span class="op">)</span></span>
<span><span class="co"># Merge the variants of GWAS and eQTL dataset by rsid:</span></span>
<span><span class="va">gwasEqtldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AF"</span><span class="op">)</span><span class="op">]</span>, <span class="va">eqtlAsso</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span>rsid<span class="op">=</span><span class="va">snpId</span>, <span class="va">pValue</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                      by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span><span class="op">)</span>, suffixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gwas"</span>,<span class="st">".eqtl"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Five retained fields are required:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwasEqtldata</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;             rsid chrom position pValue.gwas pValue.eqtl</span></span>
<span><span class="co">#&gt;    1: rs10000104  chr4 10190142      0.2830   0.4986640</span></span>
<span><span class="co">#&gt;    2: rs10000248  chr4 10720872      0.0359   0.6089460</span></span>
<span><span class="co">#&gt;    3: rs10000318  chr4 11613397      0.5810   0.3734920</span></span>
<span><span class="co">#&gt;    4: rs10000369  chr4 11234631      0.0453   0.6642180</span></span>
<span><span class="co">#&gt;    5: rs10000399  chr4 11595842      0.2370   0.0948472</span></span>
<span><span class="co">#&gt;   ---                                                  </span></span>
<span><span class="co">#&gt; 8007:  rs9999345  chr4 10895379      0.0861   0.0100352</span></span>
<span><span class="co">#&gt; 8008:  rs9999470  chr4 10315399      0.2300   0.0700527</span></span>
<span><span class="co">#&gt; 8009:  rs9999523  chr4 11410463      0.9930   0.2258890</span></span>
<span><span class="co">#&gt; 8010:  rs9999669  chr4 10852509      0.2470   0.0441530</span></span>
<span><span class="co">#&gt; 8011:  rs9999767  chr4 10127270      0.8730   0.1902930</span></span></code></pre>
<p>Visualization of p-value distribution and comparison of the signals
of GWAS and eQTL:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusCompare.html">xQTLvisual_locusCompare</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">pValue.eqtl</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                        <span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">pValue.gwas</span><span class="op">)</span><span class="op">]</span>, legend_position <span class="op">=</span> <span class="st">"bottomright"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_locuscompare.png" width="50%" style="display: block; margin: auto;"></p>
<p>Locuszoom plot of GWAS signals:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusZoom.html">xQTLvisual_locusZoom</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue.gwas</span><span class="op">)</span><span class="op">]</span>, legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_locuszoom_gwas.png" width="50%" style="display: block; margin: auto;"></p>
<p>Locuszoom plot of eQTL signals:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusZoom.html">xQTLvisual_locusZoom</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue.eqtl</span><span class="op">)</span><span class="op">]</span>, legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_locuszoom_eqtl.png" width="50%" style="display: block; margin: auto;"></p>
<p>We can also combine locuscompare and locuszoom plot using function
<code>xQTLvisual_locusCombine</code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusCombine.html">xQTLvisual_locusCombine</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>,<span class="st">"chrom"</span>, <span class="st">"position"</span>, <span class="st">"pValue.gwas"</span>, <span class="st">"pValue.eqtl"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_combined.png" width="90%" style="display: block; margin: auto;"></p>
<p>From the above figures, we can see that the SNP
<code>rs11568818</code> is potential causal variant, and we can use a
violin plot to show the normalized effect size of it:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_eqtlExp.html">xQTLvisual_eqtlExp</a></span><span class="op">(</span><span class="st">"rs11568818"</span>, <span class="st">"MMP7"</span>, tissueSiteDetail <span class="op">=</span> <span class="va">tissueSiteDetail</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_eQTL_exp.png" width="40%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="extension-integrative-analysis-with-external-packages-or-tools">Extension: Integrative analysis with external packages or tools<a class="anchor" aria-label="anchor" href="#extension-integrative-analysis-with-external-packages-or-tools"></a>
</h2>
<p>To gain insight into the function of trait genes with higher PPH4,
and explore the potential regulatory mechanism of the prostate cancer,
we can conduct exploratory analysis, like co-expression analysis, and
gene ontology enrichment analyses.</p>
<p>First we download expression profiles of the genes with higher value
of PPH4 (&gt;0.75) &amp; hypr_posterior (&gt;0.5) in prostate.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLdownload_exp.html">xQTLdownload_exp</a></span><span class="op">(</span><span class="va">outGenes</span><span class="op">$</span><span class="va">gencodeId</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span>, toSummarizedExperiment <span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Pearson coefficient can be calculated with the expression matrix for
each gene:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corDT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">expMat</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">:</span><span class="op">-</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">corDT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">outGenes</span><span class="op">$</span><span class="va">geneSymbol</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">corDT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">outGenes</span><span class="op">$</span><span class="va">geneSymbol</span></span></code></pre></div>
<p>R package <code>corrplot</code> is used to display this correlation
matrix:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/taiyun/corrplot" class="external-link">corrplot</a></span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html" class="external-link">corrplot</a></span><span class="op">(</span><span class="va">corDT</span>, method<span class="op">=</span><span class="st">"color"</span>,  </span>
<span>         type<span class="op">=</span><span class="st">"upper"</span>,    </span>
<span>         order <span class="op">=</span> <span class="st">"hclust"</span>,</span>
<span>         addCoef.col <span class="op">=</span> <span class="st">"#ff0099"</span>, </span>
<span>         number.cex <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> </span></code></pre></div>
<p><img src="images/prostate/prostate_cor.png" width="90%" style="display: block; margin: auto;"></p>
<p>R package <code>clusterProfiler</code> is used for gene functional
annotation:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">ego</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">outGenes</span><span class="op">$</span><span class="va">entrezGeneId</span><span class="op">)</span>,</span>
<span>                OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>                ont<span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                pAdjustMethod <span class="op">=</span><span class="st">"none"</span>,</span>
<span>                readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">ego</span>, showCategory<span class="op">=</span><span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/prostate/prostate_Go.png" width="80%" style="display: block; margin: auto;"></p>
<p>Viral-related GO terms are enriched in above analysis, including
<code>"viral life cycle"</code>,
<code>"positive regulation of viral life cycle"</code> and
<code>"positive regulation of viral process"</code>. Previous studies
have highlighted the role of viral infections in initiation or
progression of prostate cancer. The presence of viruses such as human
papillomavirus (HPV), herpesviruses including cytomegalovirus (CMV),
human herpes simplex virus type 2 (HSV2), human herpesvirus type 8
(HHV8) and Epstein-Barr virus (EBV) can infect the prostate <span class="citation">(<a href="#ref-abidi2018" role="doc-biblioref">Abidi et
al. 2018</a>)</span>. However, the causal variants’ genetic effects on
the phenotype and whether the trait genes has a direct association with
prostate carcinogenesis has not yet been established.</p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-abidi2018" class="csl-entry">
Abidi, Syed Hani, Fareena Bilwani, Kulsoom Ghias, and Farhat Abbas.
2018. <span>“Viral etiology of prostate cancer: Genetic alterations and
immune response. A literature review.”</span> <em>International Journal
of Surgery (London, England)</em> 52 (April): 136–40. <a href="https://doi.org/10.1016/j.ijsu.2018.02.050" class="external-link">https://doi.org/10.1016/j.ijsu.2018.02.050</a>.
</div>
<div id="ref-RefSeq" class="csl-entry">
<span>“RefSeq.”</span> n.d. <a href="https://www.ncbi.nlm.nih.gov/gene/4316#summary" class="external-link">https://www.ncbi.nlm.nih.gov/gene/4316#summary</a>.
</div>
<div id="ref-tregunna2020" class="csl-entry">
Tregunna, Rebecca. 2020. <span>“Serum MMP7 levels could guide metastatic
therapy for prostate cancer.”</span> <em>Nature Reviews. Urology</em> 17
(12): 658. <a href="https://doi.org/10.1038/s41585-020-00396-3" class="external-link">https://doi.org/10.1038/s41585-020-00396-3</a>.
</div>
<div id="ref-zhang2017" class="csl-entry">
Zhang, Q, S Liu, K R Parajuli, W Zhang, K Zhang, Z Mo, J Liu, et al.
2017. <span>“Interleukin-17 Promotes Prostate Cancer via MMP7-Induced
Epithelial-to-Mesenchymal Transition.”</span> <em>Oncogene</em> 36 (5):
687–99. <a href="https://doi.org/10.1038/onc.2016.240" class="external-link">https://doi.org/10.1038/onc.2016.240</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ruofan Ding, Xudong Zou, Yangmei Qin, Gao Wang, Lei Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
