<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A quick application of coloclization analysis • xQTLbiolinks</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A quick application of coloclization analysis">
<meta property="og:description" content="xQTLbiolinks">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xQTLbiolinks</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="https://github.com/dingruofan/xQTLbiolinks/wiki/Visualization-of-expression-and-xQTL" class="external-link">Visualization</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A quick application of coloclization analysis</h1>
            
            <h4 data-toc-skip class="date">2022-06-08</h4>
      
      
      <div class="hidden name"><code>Quick_start.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">xQTLbiolinks</span><span class="op">)</span></code></pre></div>
<pre><code>&gt; xQTLbiolinks v0.1.0  For help: https://github.com/lilab-bioinfo/xQTLbiolinks/
&gt; 
&gt; If you use xQTLbiolinks in published research, please cite:
&gt; Ruofan Ding, Xudong Zou, Gao Wang, Lei Li. xQTLbiolinks: an R/Bioconductor package for integrative analysis of xQTL data.</code></pre>
<div class="section level2">
<h2 id="a-quick-application-of-coloclization-analysis">A quick application of coloclization analysis<a class="anchor" aria-label="anchor" href="#a-quick-application-of-coloclization-analysis"></a>
</h2>
<div class="section level3">
<h3 id="before-we-start">Before we start<a class="anchor" aria-label="anchor" href="#before-we-start"></a>
</h3>
<p>Colocalization analysis can be performed with <code>xQTLbiolinks</code> by providing their own GWAS summary statistics data. All we need to prepare include three parts:</p>
<ol style="list-style-type: decimal">
<li>GWAS summary statistics dataset.</li>
<li>Name of tissue of interest.</li>
<li>Following three installed packages.</li>
</ol>
</div>
<div class="section level3">
<h3 id="step-1--data-preparation-">Step 1. data preparation.<a class="anchor" aria-label="anchor" href="#step-1--data-preparation-"></a>
</h3>
<p>Download and load an example file of summary statistics dataset (GRCh38). We perform colocalization analysis in <code>Brain - Cerebellum</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gwasDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"http://raw.githubusercontent.com/dingruofan/exampleData/master/gwas/AD/gwasDFsub.txt"</span><span class="op">)</span>
<span class="va">tissueSiteDetail</span><span class="op">=</span><span class="st">"Brain - Cerebellum"</span></code></pre></div>
<p>In this example, a data.table object of 16538 (rows) x 5 (cols) is loaded. Five columns are required (arbitrary column names is supported, but columns must be as the following order):</p>
<p><code>Col 1</code>. “variants” (character), , using an rsID (e.g. “rs11966562”);</p>
<p><code>Col 2</code>. “chromosome” (character), one of the chromosome from chr1-chr22;</p>
<p><code>Col 3</code>. “position” (integer), genome position of snp;</p>
<p><code>Col 4</code>. “P-value” (numeric);</p>
<p><code>Col 5</code>. “MAF” (numeric). Allel frequency;</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">rsid</th>
<th align="left">chr</th>
<th align="left">position</th>
<th>P</th>
<th>maf</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rs13120565</td>
<td align="left">chr4</td>
<td align="left">10702513</td>
<td>5.66e-10</td>
<td>0.6429</td>
</tr>
<tr class="even">
<td align="left">rs4697781</td>
<td align="left">chr4</td>
<td align="left">10703243</td>
<td>8.94e-10</td>
<td>0.6446</td>
</tr>
<tr class="odd">
<td align="left">rs4697779</td>
<td align="left">chr4</td>
<td align="left">10701187</td>
<td>5.74e-09</td>
<td>0.3231</td>
</tr>
<tr class="even">
<td align="left">rs4697780</td>
<td align="left">chr4</td>
<td align="left">10701381</td>
<td>5.95e-09</td>
<td>0.3231</td>
</tr>
<tr class="odd">
<td align="left">rs4547789</td>
<td align="left">chr4</td>
<td align="left">10702891</td>
<td>1.46e-08</td>
<td>0.3197</td>
</tr>
<tr class="even">
<td align="left">rs11726285</td>
<td align="left">chr4</td>
<td align="left">10700944</td>
<td>1.47e-08</td>
<td>0.3214</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="step-2--filter-sentinel-snps-">Step 2. Filter sentinel snps.<a class="anchor" aria-label="anchor" href="#step-2--filter-sentinel-snps-"></a>
</h3>
<p>Sentinel SNPs can be detected using <code>xQTLanalyze_getSentinelSnp</code> with the arguments <code>p-value &lt; 5e-8</code> and <code>SNP-to-SNP distance &gt; 10e6 bp</code>. We recommend converting the genome version of the GWAS file to GRCh38 if that is GRCh37 (run with parameter: <code>genomeVersion="grch37"</code>; <code>grch37To38=TRUE</code>, and package <code>rtracklayer</code>is required).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sentinelSnpDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_getSentinelSnp.html">xQTLanalyze_getSentinelSnp</a></span><span class="op">(</span><span class="va">gwasDF</span>, pValueThreshold <span class="op">=</span> <span class="fl">5e-08</span><span class="op">)</span></code></pre></div>
<p>After filtering, a sentinel SNP with P-value&lt;5e-8 is detected in this example:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sentinelSnpDF</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th>rsid</th>
<th>chr</th>
<th>position</th>
<th>pValue</th>
<th>maf</th>
</tr></thead>
<tbody><tr class="odd">
<td>rs13120565</td>
<td>chr4</td>
<td>10702513</td>
<td>5.66e-10</td>
<td>0.6429</td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="step-3--identify-trait-genes-for-each-sentinel-snps-">Step 3. Identify trait genes for each sentinel SNPs.<a class="anchor" aria-label="anchor" href="#step-3--identify-trait-genes-for-each-sentinel-snps-"></a>
</h3>
<p>Trait genes are genes that located in the range of 1Mb (default, can be changed with parameter <code>detectRange</code>) of sentinel SNPs. Besides, In order to reduce the number of trait genes and thus reduce the running time, we take the overlap of eGenes and trait genes as the final output of the function <code>xQTLanalyze_getTraits</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">traitsAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_getTraits.html">xQTLanalyze_getTraits</a></span><span class="op">(</span><span class="va">sentinelSnpDF</span>, detectRange<span class="op">=</span><span class="fl">1e6</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span></code></pre></div>
<p>Totally, 3 associations between 3 traits genes and 1 SNPs are detected</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">traitsAll</span></code></pre></div>
<table class="table">
<colgroup>
<col width="10%">
<col width="9%">
<col width="8%">
<col width="10%">
<col width="12%">
<col width="16%">
<col width="10%">
<col width="8%">
<col width="8%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th>chromosome</th>
<th>geneStart</th>
<th>geneEnd</th>
<th>geneStrand</th>
<th>geneSymbol</th>
<th>gencodeId</th>
<th>rsid</th>
<th>position</th>
<th>pValue</th>
<th>maf</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>chr4</td>
<td>11393150</td>
<td>11429765</td>
<td>-</td>
<td>HS3ST1</td>
<td>ENSG00000002587.9</td>
<td>rs13120565</td>
<td>10702513</td>
<td>5.66e-10</td>
<td>0.6429</td>
</tr>
<tr class="even">
<td>chr4</td>
<td>10486395</td>
<td>10684865</td>
<td>-</td>
<td>CLNK</td>
<td>ENSG00000109684.14</td>
<td>rs13120565</td>
<td>10702513</td>
<td>5.66e-10</td>
<td>0.6429</td>
</tr>
<tr class="odd">
<td>chr4</td>
<td>10068089</td>
<td>10073019</td>
<td>-</td>
<td>RP11-448G15.3</td>
<td>ENSG00000261490.1</td>
<td>rs13120565</td>
<td>10702513</td>
<td>5.66e-10</td>
<td>0.6429</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="step-4--conduct-colocalization-analysis-">Step 4. Conduct colocalization analysis.<a class="anchor" aria-label="anchor" href="#step-4--conduct-colocalization-analysis-"></a>
</h3>
<p>Following three steps of colocalization analysis are encapsulated in one function <code>xQTLanalyze_coloc</code>:</p>
<ul>
<li>Retrieved all associations from EBI eQTL catalogue for a specified gene.</li>
<li>Merge the data.frame of GWAS and eQTL by rsid.</li>
<li>Perform colocalization analysis using<code>coloc</code> package.</li>
</ul>
<p><strong>For a single trait gene</strong>, like <em>CLNK</em> in above table, colocalization analysis can be performed with:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_coloc.html">xQTLanalyze_coloc</a></span><span class="op">(</span><span class="va">gwasDF</span>, <span class="st">"CLNK"</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span> <span class="co"># using gene symbol</span></code></pre></div>
<p><em>output</em> is a list, including three parts: <em>coloc_Out_summary</em>, <em>gwasEqtlInfo</em>, and <em>gwasEqtlInfo</em>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">output</span><span class="op">$</span><span class="va">coloc_Out_summary</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th>nsnps</th>
<th>PP.H0.abf</th>
<th>PP.H1.abf</th>
<th>PP.H2.abf</th>
<th>PP.H3.abf</th>
<th>PP.H4.abf</th>
<th>traitGene</th>
</tr></thead>
<tbody><tr class="odd">
<td>7108</td>
<td>7.097893e-11</td>
<td>1.32221e-07</td>
<td>3.890211e-06</td>
<td>0.00625302</td>
<td>0.993743</td>
<td>CLNK</td>
</tr></tbody>
</table>
<p><strong>For multiple trait genes</strong>, a for loop or lapply function can be used to get all genes’ outputs.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">outputs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html" class="external-link">rbindlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">traitsAll</span><span class="op">$</span><span class="va">gencodeId</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span> <span class="co"># using gencode ID.</span>
 <span class="fu"><a href="../reference/xQTLanalyze_coloc.html">xQTLanalyze_coloc</a></span><span class="op">(</span><span class="va">gwasDF</span>, <span class="va">x</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span><span class="op">$</span><span class="va">coloc_Out_summary</span> <span class="op">}</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><em>outputs</em> is a data.table that combined all results of <em>coloc_Out_summary</em> of all genes.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">outputs</span></code></pre></div>
<table class="table">
<colgroup>
<col width="9%">
<col width="14%">
<col width="14%">
<col width="14%">
<col width="12%">
<col width="12%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th>nsnps</th>
<th>PP.H0.abf</th>
<th>PP.H1.abf</th>
<th>PP.H2.abf</th>
<th>PP.H3.abf</th>
<th>PP.H4.abf</th>
<th>traitGene</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>6453</td>
<td>1.730172e-05</td>
<td>3.218423e-02</td>
<td>6.603463e-05</td>
<td>0.12199027</td>
<td>0.84574216</td>
<td>ENSG00000002587.9</td>
</tr>
<tr class="even">
<td>7108</td>
<td>7.097893e-11</td>
<td>1.322210e-07</td>
<td>3.890211e-06</td>
<td>0.00625302</td>
<td>0.99374296</td>
<td>ENSG00000109684.14</td>
</tr>
<tr class="odd">
<td>6601</td>
<td>5.287051e-05</td>
<td>9.848309e-02</td>
<td>4.801374e-04</td>
<td>0.89435622</td>
<td>0.00662768</td>
<td>ENSG00000261490.1</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="step-5--visualization-of-the-results-">Step 5. Visualization of the results.<a class="anchor" aria-label="anchor" href="#step-5--visualization-of-the-results-"></a>
</h3>
<p>For the potential casual gene (<strong><em>ENSG00000187323.11</em></strong>, PP4=<strong>0.9937</strong>), We merge the variants of GWAS and eQTL by rsid.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">eqtlAsso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLdownload_eqtlAllAsso.html">xQTLdownload_eqtlAllAsso</a></span><span class="op">(</span>gene<span class="op">=</span><span class="st">"ENSG00000109684.14"</span>, 
                                     tissueSiteDetail <span class="op">=</span> <span class="va">tissueSiteDetail</span>, 
                                     withB37VariantId<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">gwasEqtldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">gwasDF</span>, <span class="va">eqtlAsso</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span>rsid<span class="op">=</span><span class="va">snpId</span>, position<span class="op">=</span><span class="va">pos</span>, <span class="va">maf</span>, <span class="va">pValue</span><span class="op">)</span><span class="op">]</span>,
                      by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>, <span class="st">"position"</span><span class="op">)</span>, suffixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gwas"</span>,<span class="st">".eqtl"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Visualization of p-value distribution and comparison of the signals of GWAS and eQTL:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/xQTLvisual_locusCompare.html">xQTLvisual_locusCompare</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">pValue.eqtl</span><span class="op">)</span><span class="op">]</span>, 
                       <span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">pValue.gwas</span><span class="op">)</span><span class="op">]</span>, legend_position <span class="op">=</span> <span class="st">"bottomright"</span><span class="op">)</span></code></pre></div>
<p><img src="Quick_start_files/figure-html/unnamed-chunk-15-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Locuszoom plot of GWAS signals:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/xQTLvisual_locusZoom.html">xQTLvisual_locusZoom</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue.gwas</span><span class="op">)</span><span class="op">]</span>, legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="Quick_start_files/figure-html/unnamed-chunk-17-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Locuszoom plot of eQTL signals:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/xQTLvisual_locusZoom.html">xQTLvisual_locusZoom</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue.eqtl</span><span class="op">)</span><span class="op">]</span>, 
                     highlightSnp <span class="op">=</span> <span class="st">"rs13120565"</span>, legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="Quick_start_files/figure-html/unnamed-chunk-19-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Violin plot of normalized exprssion of eQTL:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/xQTLvisual_eqtlExp.html">xQTLvisual_eqtlExp</a></span><span class="op">(</span><span class="st">"rs13120565"</span>, <span class="st">"ENSG00000109684.14"</span>, tissueSiteDetail <span class="op">=</span> <span class="va">tissueSiteDetail</span><span class="op">)</span></code></pre></div>
<p><img src="Quick_start_files/figure-html/unnamed-chunk-21-1.png" width="316.8" style="display: block; margin: auto;"></p>
<p>We can also combine locuscompare and locuszoom plot using <code>xQTLvisual_locusCombine</code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/xQTLvisual_locusCombine.html">xQTLvisual_locusCombine</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>,<span class="st">"chrom"</span>, <span class="st">"position"</span>, <span class="st">"pValue.gwas"</span>, <span class="st">"pValue.eqtl"</span><span class="op">)</span><span class="op">]</span>, 
                        highlightSnp<span class="op">=</span><span class="st">"rs13120565"</span><span class="op">)</span></code></pre></div>
<p><img src="Quick_start_files/figure-html/unnamed-chunk-23-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ruofan Ding.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
