<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2. A quick application of coloclization analysis • xQTLbiolinks</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2. A quick application of coloclization analysis">
<meta property="og:description" content="xQTLbiolinks">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xQTLbiolinks</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/visualization.html">Visualization</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dingruofan/xQTLbiolinks/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2. A quick application of coloclization
analysis</h1>
            
            <h4 data-toc-skip class="date">2023-05-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dingruofan/xQTLbiolinks/blob/HEAD/vignettes/Quick_start.Rmd" class="external-link"><code>vignettes/Quick_start.Rmd</code></a></small>
      <div class="hidden name"><code>Quick_start.Rmd</code></div>

    </div>

    
    
<p>xQTLbiolinks has provided a pipeline for disease target gene
discovery through colocalization analysis with xQTLs to detect
susceptible genes and causal variants.
<img src="images/quick_start/colocalization_framework.png" width="70%" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="before-we-start">Before we start<a class="anchor" aria-label="anchor" href="#before-we-start"></a>
</h2>
<p>To perform colocalization analysis using xQTLbiolinks, users need to
provide their own <strong>GWAS summary statistics data</strong>, specify
the <strong>tissue</strong> of interest and import following
packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dingruofan/xQTLbiolinks" class="external-link">xQTLbiolinks</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chr1swallace/coloc" class="external-link">coloc</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-1--data-pre-processing">Step 1. data pre-processing<a class="anchor" aria-label="anchor" href="#step-1--data-pre-processing"></a>
</h2>
<p>Download and load an example file of summary statistics dataset
(GRCh38). We perform colocalization analysis in
<code>Brain - Cerebellum</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwasDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"https://master.dl.sourceforge.net/project/exampledata/gwasDFsub.txt"</span><span class="op">)</span></span>
<span><span class="va">tissueSiteDetail</span><span class="op">=</span><span class="st">"Brain - Cerebellum"</span></span></code></pre></div>
<p>In this example, a data.table object of 16538 (rows) x 5 (cols) is
loaded. Five columns are required (arbitrary column names is supported,
but columns must be as the following order):</p>
<p><code>Col 1</code>. “variants” (character), , using an rsID
(e.g. “rs11966562”);</p>
<p><code>Col 2</code>. “chromosome” (character), one of the chromosome
from chr1-chr22;</p>
<p><code>Col 3</code>. “position” (integer), genome position of snp;</p>
<p><code>Col 4</code>. “P-value” (numeric);</p>
<p><code>Col 5</code>. “MAF” (numeric). Allel frequency;</p>
<p><code>Col 6</code>. “beta” (numeric). effect size.</p>
<p><code>Col 7</code>. “se” (numeric). standard error.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwasDF</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;          rsid chrom position   pValue     AF     beta      se</span></span>
<span><span class="co">#&gt; 1: rs13120565     4 10702513 5.66e-10 0.6429  0.01825 0.00294</span></span>
<span><span class="co">#&gt; 2:  rs4697781     4 10703243 8.94e-10 0.6446  0.01803 0.00294</span></span>
<span><span class="co">#&gt; 3:  rs4697779     4 10701187 5.74e-09 0.3231 -0.01747 0.00300</span></span>
<span><span class="co">#&gt; 4:  rs4697780     4 10701381 5.95e-09 0.3231 -0.01746 0.00300</span></span>
<span><span class="co">#&gt; 5:  rs4547789     4 10702891 1.46e-08 0.3197 -0.01703 0.00300</span></span>
<span><span class="co">#&gt; 6: rs11726285     4 10700944 1.47e-08 0.3214 -0.01702 0.00301</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-2--identify-sentinel-snps-">Step 2. Identify sentinel snps.<a class="anchor" aria-label="anchor" href="#step-2--identify-sentinel-snps-"></a>
</h2>
<p>Sentinel SNPs can be detected using
<code>xQTLanalyze_getSentinelSnp</code> with the arguments
<code>p-value &lt; 5e-8</code> and
<code>SNP-to-SNP distance &gt; 10e6 bp</code>. We recommend converting
the genome version of the GWAS file to GRCh38 if that is GRCh37 (run
with parameter: <code>genomeVersion="grch37"</code>;
<code>grch37To38=TRUE</code>, and package <code>rtracklayer</code>is
required).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sentinelSnpDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_getSentinelSnp.html">xQTLanalyze_getSentinelSnp</a></span><span class="op">(</span><span class="va">gwasDF</span>, pValueThreshold <span class="op">=</span> <span class="fl">5e-08</span><span class="op">)</span></span></code></pre></div>
<p>After filtering, a sentinel SNP with P-value&lt;5e-8 is detected in
this example:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sentinelSnpDF</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;          rsid  chr position   pValue    maf    beta      se</span></span>
<span><span class="co">#&gt; 1: rs13120565 chr4 10702513 5.66e-10 0.6429 0.01825 0.00294</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-3--identify-trait-genes-for-each-sentinel-snps-">Step 3. Identify trait genes for each sentinel SNPs.<a class="anchor" aria-label="anchor" href="#step-3--identify-trait-genes-for-each-sentinel-snps-"></a>
</h2>
<p>Trait genes are genes that located in the range of 1Mb (default, can
be changed with parameter <code>detectRange</code>) of sentinel SNPs.
Every gene within 1Mb of sentinel SNPs is searched by fuction
<code>xQTLanalyze_getTraits</code>. Besides, In order to reduce the
number of trait genes and thus reduce the running time, we take the
overlap of eGenes and trait genes as the final output of the function
<code>xQTLanalyze_getTraits</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traitsAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_getTraits.html">xQTLanalyze_getTraits</a></span><span class="op">(</span><span class="va">sentinelSnpDF</span>, detectRange<span class="op">=</span><span class="fl">1e6</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span></span></code></pre></div>
<p>Totally, 3 associations between 3 traits genes and 1 SNPs are
detected</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traitsAll</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;          rsid  chr position   pValue    maf    beta      se       gencodeId</span></span>
<span><span class="co">#&gt; 1: rs13120565 chr4 10702513 5.66e-10 0.6429 0.01825 0.00294 ENSG00000002587</span></span>
<span><span class="co">#&gt; 2: rs13120565 chr4 10702513 5.66e-10 0.6429 0.01825 0.00294 ENSG00000109684</span></span>
<span><span class="co">#&gt; 3: rs13120565 chr4 10702513 5.66e-10 0.6429 0.01825 0.00294 ENSG00000261490</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-4--conduct-colocalization-analysis-">Step 4. Conduct colocalization analysis.<a class="anchor" aria-label="anchor" href="#step-4--conduct-colocalization-analysis-"></a>
</h2>
<p><strong>For a single trait gene</strong>, like
<em>ENSG00000109684</em> in above table, colocalization analysis (using
<code>coloc</code> method) can be performed with:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLanalyze_coloc.html">xQTLanalyze_coloc</a></span><span class="op">(</span><span class="va">gwasDF</span>, <span class="st">"ENSG00000109684"</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span><span class="op">)</span> <span class="co"># using gene symbol</span></span></code></pre></div>
<p><em>output</em> is a list, including two parts:
<em>coloc_Out_summary</em> and <em>gwasEqtlInfo</em>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span><span class="op">$</span><span class="va">coloc_Out_summary</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;    nsnps    PP.H0.abf   PP.H1.abf    PP.H2.abf  PP.H3.abf PP.H4.abf</span></span>
<span><span class="co">#&gt; 1:  7107 7.097893e-11 1.32221e-07 3.890211e-06 0.00625302  0.993743</span></span>
<span><span class="co">#&gt;          traitGene candidate_snp SNP.PP.H4</span></span>
<span><span class="co">#&gt; 1: ENSG00000109684    rs13120565 0.5328849</span></span></code></pre>
<p><strong>For multiple trait genes</strong>, a for loop or lapply
function can be used to get all genes’ outputs (using both
<code>coloc</code> and <code>hyprcoloc</code> methods).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html" class="external-link">rbindlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">traitsAll</span><span class="op">$</span><span class="va">gencodeId</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span> <span class="co"># using gencode ID.</span></span>
<span>           <span class="fu"><a href="../reference/xQTLanalyze_coloc.html">xQTLanalyze_coloc</a></span><span class="op">(</span><span class="va">gwasDF</span>, <span class="va">x</span>, tissueSiteDetail<span class="op">=</span><span class="va">tissueSiteDetail</span>, method <span class="op">=</span> <span class="st">"Both"</span><span class="op">)</span><span class="op">$</span><span class="va">colocOut</span> <span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><em>outputs</em> is a data.table that combined all results of
<em>coloc_Out_summary</em> of all genes.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputs</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;          traitGene nsnps    PP.H0.abf    PP.H1.abf    PP.H2.abf  PP.H3.abf</span></span>
<span><span class="co">#&gt; 1: ENSG00000002587  6452 1.730175e-05 3.218430e-02 6.603361e-05 0.12198838</span></span>
<span><span class="co">#&gt; 2: ENSG00000109684  7107 7.097893e-11 1.322210e-07 3.890211e-06 0.00625302</span></span>
<span><span class="co">#&gt; 3: ENSG00000261490  6601 5.287051e-05 9.848309e-02 4.801374e-04 0.89435622</span></span>
<span><span class="co">#&gt;     PP.H4.abf candidate_snp SNP.PP.H4 hypr_posterior hypr_regional_prob</span></span>
<span><span class="co">#&gt; 1: 0.84574398    rs13120565 0.4140146         0.5685             0.9694</span></span>
<span><span class="co">#&gt; 2: 0.99374296    rs13120565 0.5328849         0.9793             0.9999</span></span>
<span><span class="co">#&gt; 3: 0.00662768    rs13120565 0.4219650         0.0000             0.0101</span></span>
<span><span class="co">#&gt;    hypr_candidate_snp hypr_posterior_explainedBySnp</span></span>
<span><span class="co">#&gt; 1:         rs13120565                        0.2726</span></span>
<span><span class="co">#&gt; 2:         rs13120565                        0.4747</span></span>
<span><span class="co">#&gt; 3:         rs13120565                        0.4118</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-5--visualization-of-the-results-">Step 5. Visualization of the results.<a class="anchor" aria-label="anchor" href="#step-5--visualization-of-the-results-"></a>
</h2>
<p>For the potential casual gene
<strong><em>ENSG00000109684</em></strong> (PP4=<strong>0.9937</strong>
&amp; hypr_posterior=<strong>0.9793</strong>), we can get its
significant associations across tissues:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_eqtl.html">xQTLvisual_eqtl</a></span><span class="op">(</span><span class="st">"ENSG00000109684"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/quick_start/eQTL_multiple_tissue.png" width="80%" style="display: block; margin: auto;"></p>
<p>For the purpose of visualization of p-value distribution and
comparison of the signals of GWAS and eQTL: We merge the variants of
GWAS and eQTL by rsid first.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eqtlAsso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLdownload_eqtlAllAsso.html">xQTLdownload_eqtlAllAsso</a></span><span class="op">(</span>gene<span class="op">=</span><span class="st">"ENSG00000109684"</span>, </span>
<span>                                     tissueLabel <span class="op">=</span> <span class="va">tissueSiteDetail</span><span class="op">)</span></span>
<span><span class="va">gwasEqtldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">gwasDF</span>, <span class="va">eqtlAsso</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span>rsid<span class="op">=</span><span class="va">snpId</span>, position<span class="op">=</span><span class="va">pos</span>, <span class="va">maf</span>, <span class="va">pValue</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                      by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>, <span class="st">"position"</span><span class="op">)</span>, suffixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".gwas"</span>,<span class="st">".eqtl"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Function <code>xQTLvisual_locusCompare</code> displays the candidate
SNP <strong><em>rs13120565</em></strong>
(SNP.PP.H4=<strong>0.5328849</strong> &amp;
hypr_posterior_explainedBySnp=<strong>0.4747</strong>) in the top right
corner:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusCompare.html">xQTLvisual_locusCompare</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">pValue.eqtl</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                        <span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">pValue.gwas</span><span class="op">)</span><span class="op">]</span>, legend_position <span class="op">=</span> <span class="st">"bottomright"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note: </strong>The information on SNP linkage disequilibrium
is automatically downloaded online. Due to network issues, the download
may fail at times. If this happens, please try running it again.</p>
<p><img src="images/quick_start/compare.png" width="50%" style="display: block; margin: auto;"></p>
<p>Locuszoom plot of GWAS signals:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusZoom.html">xQTLvisual_locusZoom</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue.gwas</span><span class="op">)</span><span class="op">]</span>, legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/quick_start/gwas_legendF.png" width="50%" style="display: block; margin: auto;"></p>
<p>LocusZoom plot of eQTL signals:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusZoom.html">xQTLvisual_locusZoom</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">rsid</span>, <span class="va">chrom</span>, <span class="va">position</span>, <span class="va">pValue.eqtl</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                     highlightSnp <span class="op">=</span> <span class="st">"rs13120565"</span>, legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/quick_start/eqtl_legendF.png" width="50%" style="display: block; margin: auto;"></p>
<p>Violin plot of normalized expression of eQTL
(rs13120565-ENSG00000187323.11):</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_eqtlExp.html">xQTLvisual_eqtlExp</a></span><span class="op">(</span><span class="st">"rs13120565"</span>, <span class="st">"ENSG00000109684"</span>, tissueSiteDetail <span class="op">=</span> <span class="va">tissueSiteDetail</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/quick_start/exp.png" width="45%" style="display: block; margin: auto;"></p>
<p>We can also combine locuscompare and locuszoom plot using
<code>xQTLvisual_locusCombine</code>:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xQTLvisual_locusCombine.html">xQTLvisual_locusCombine</a></span><span class="op">(</span><span class="va">gwasEqtldata</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>,<span class="st">"chrom"</span>, <span class="st">"position"</span>, <span class="st">"pValue.gwas"</span>, <span class="st">"pValue.eqtl"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                        highlightSnp<span class="op">=</span><span class="st">"rs13120565"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/quick_start/xQTLvisual_locusCombine.png" width="90%" style="display: block; margin: auto;"></p>
<p>Colocalized loci should show a general pattern where SNPs in high LD
will show strong associations with expression levels of the colocalized
gene, and the eQTL associations will weaken for SNPs in lower LD. This
pattern of the eQTL varies among different tissues / cell-types, the
strength of which indicates the power of the regulatory effect of the
variant. We can visualize the correlation between p-values of eQTLs and
LD across numerous tissues/cell types to easily distinguish this patten
using <code>xQTLvisual_coloc</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multi_tissue_coloc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xQTLvisual_coloc.html">xQTLvisual_coloc</a></span><span class="op">(</span>gene<span class="op">=</span><span class="st">"ENSG00000109684"</span>, variantName<span class="op">=</span><span class="st">"rs13120565"</span>, </span>
<span>                                      tissueLabels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Brain - Cerebellar Hemisphere"</span>, </span>
<span>                                                        <span class="st">"Brain - Cerebellum"</span>, <span class="st">"Thyroid"</span>, <span class="st">"Lung"</span>,</span>
<span>                                                        <span class="st">"Cells - EBV-transformed lymphocytes"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/quick_start/propensity.png" width="70%" style="display: block; margin: auto;"></p>
<p>For applying <code>xQTLbiolinks</code> to a whole case study, please
find this <a href="https://dingruofan.github.io/xQTLbiolinks/articles/Colocalization_analysis_with_xQTLbiolinks.html" class="external-link"><strong>Document</strong></a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ruofan Ding, Xudong Zou, Yangmei Qin, Gao Wang, Lei Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
