<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="users can use this function to search gene of interest"><title>Query gene information through all genes' information — apiRef_genes • xQTLbiolinks</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Query gene information through all genes' information — apiRef_genes"><meta property="og:description" content="users can use this function to search gene of interest"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">xQTLbiolinks</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Query gene information through all genes' information</h1>
      
      <div class="d-none name"><code>apiRef_genes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>users can use this function to search gene of interest</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">apiRef_genes</span><span class="op">(</span>
  genes <span class="op">=</span> <span class="st">""</span>,
  geneType <span class="op">=</span> <span class="st">"geneSymbol"</span>,
  gencodeVersion <span class="op">=</span> <span class="st">"v26"</span>,
  genomeBuild <span class="op">=</span> <span class="st">"GRCh38/hg38"</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>genes</dt>
<dd><p>Following gene types are supported:</p><ul><li><p><strong>Gene symbol</strong>.</p>
<p>A character string or a character vector (case ignored). like: "tp53","naDK","SDF4".</p></li>
<li><p><strong>Gencode/ensemble id</strong> (versioned or unversioned).</p>
<p>A character string or a character vector (case ignored). like: "ENSG00000210195.2","ENSG00000078808".</p></li>
<li><p><strong>Entrez gene ID</strong>.</p>
<p>A integer string or a integer vectors. like: 51150,5590,4509.</p></li>
<li><p><strong>geneCategory</strong>.</p>
<p>When choose "geneCategory", "genes" must be chosen from following gene category:</p><ul><li><p>protein coding</p></li>
<li><p>antisense</p></li>
<li><p>lincRNA</p></li>
<li><p>unprocessed pseudogene</p></li>
<li><p>miRNA</p></li>
<li><p>transcribed unprocessed pseudogene</p></li>
<li><p>snRNA</p></li>
<li><p>processed pseudogene</p></li>
<li><p>processed transcript</p></li>
<li><p>TEC</p></li>
<li><p>transcribed unitary pseudogene</p></li>
<li><p>transcribed processed pseudogene</p></li>
<li><p>sense intronic</p></li>
<li><p>misc RNA</p></li>
<li><p>snoRNA</p></li>
<li><p>scaRNA</p></li>
<li><p>rRNA</p></li>
<li><p>unitary pseudogene</p></li>
<li><p>3prime overlapping ncRNA</p></li>
<li><p>polymorphic pseudogene</p></li>
<li><p>bidirectional promoter lncRNA</p></li>
<li><p>sense overlapping</p></li>
<li><p>pseudogene</p></li>
<li><p>IG V pseudogene</p></li>
<li><p>scRNA</p></li>
<li><p>IG C gene</p></li>
<li><p>IG J gene</p></li>
<li><p>IG V gene</p></li>
<li><p>sRNA</p></li>
<li><p>ribozyme</p></li>
<li><p>vaultRNA</p></li>
<li><p>non coding</p></li>
<li><p>TR J gene</p></li>
<li><p>TR C gene</p></li>
<li><p>TR V gene</p></li>
<li><p>TR V pseudogene</p></li>
<li><p>TR D gene</p></li>
<li><p>IG C pseudogene</p></li>
<li><p>macro lncRNA</p></li>
<li><p>TR J pseudogene</p></li>
<li><p>IG D gene</p></li>
<li><p>IG J pseudogene</p></li>
<li><p>IG pseudogene</p></li>
<li><p>Mt tRNA</p></li>
<li><p>Mt rRNA</p></li>
</ul></li>
</ul></dd>
<dt>geneType</dt>
<dd><p>A character string. Types of queried genes. Options: "geneSymbol" (default), "gencodeId", "entrezId";</p></dd>
<dt>gencodeVersion</dt>
<dd><p>A character string. Two version are supported: "v26" (default) and "v19"</p></dd>
<dt>genomeBuild</dt>
<dd><p>A character string. "GRCh38/hg38"(default) for gencodeVersion "v26", "GRCh37/hg19" for gencodeVersion "v19"</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.table of queried gene information. including following columns:</p><ul><li><p><strong>genes.</strong> Input genes</p></li>
<li><p><strong>geneSymbol.</strong> Gene symbol.</p></li>
<li><p><strong>gencodeId.</strong> Gencode/ensemble id (versioned).</p></li>
<li><p><strong>entrezGeneId.</strong> Entrez gene ID.</p></li>
<li><p><strong>geneType.</strong> Gene type.</p></li>
<li><p><strong>chromosome.</strong> Note: "chr" is added in gencode v26,</p></li>
<li><p><strong>start.</strong></p></li>
<li><p><strong>end.</strong></p></li>
<li><p><strong>strand.</strong></p></li>
<li><p><strong>tss.</strong> Transcriptional start site.</p></li>
<li><p><strong>gencodeVersion.</strong> Gencode Version.</p></li>
<li><p><strong>genomeBuild.</strong> Genome version.</p></li>
<li><p><strong>description.</strong></p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Ruofan Ding.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

