<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ xQTLbiolinks</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xQTLbiolinks</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/visualization.html">Visualization</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dingruofan/xQTLbiolinks/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="data-retrieval">Data retrieval <a href="#data-retrieval" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="xQTLquery_eqtl.html">xQTLquery_eqtl()</a></code> </p>
        </td>
        <td><p>Query multi-tissue eQTL metasoft results</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_eqtlSig.html">xQTLquery_eqtlSig()</a></code> </p>
        </td>
        <td><p>Query significant eQTL associations for a specified tissue or multiple tissues.</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_gene.html">xQTLquery_gene()</a></code> </p>
        </td>
        <td><p>Query basic information for genes, including name, symbol, position and description</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_geneAll.html">xQTLquery_geneAll()</a></code> </p>
        </td>
        <td><p>Query all genes supported in GTEx</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_sampleByTissue.html">xQTLquery_sampleByTissue()</a></code> </p>
        </td>
        <td><p>Query details of samples by tissue name</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_sc.html">xQTLquery_sc()</a></code> </p>
        </td>
        <td><p>Query significant sc-eQTLs for a specified gene</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_scInfo.html">xQTLquery_scInfo()</a></code> </p>
        </td>
        <td><p>Query metadata of sc-eQTLs</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_sqtlSig.html">xQTLquery_sqtlSig()</a></code> </p>
        </td>
        <td><p>Query significant sQTL associations for a tissue or multple tissues</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_tissue.html">xQTLquery_tissue()</a></code> </p>
        </td>
        <td><p>Query details for a specified tissue</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_varId.html">xQTLquery_varId()</a></code> </p>
        </td>
        <td><p>Query variant with variant ID or dbSNP ID</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLquery_varPos.html">xQTLquery_varPos()</a></code> </p>
        </td>
        <td><p>Query variants using genome position.</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_egene.html">xQTLdownload_egene()</a></code> </p>
        </td>
        <td><p>Download eGenes (eQTL Genes) for a specified gene or a tissue</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_eqtlAllAsso.html">xQTLdownload_eqtlAllAsso()</a></code> </p>
        </td>
        <td><p>Download summary statistics data for eQTLs with a specified gene, variant, tissue or study</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_eqtlAllAssoPos.html">xQTLdownload_eqtlAllAssoPos()</a></code> </p>
        </td>
        <td><p>Download summary statistics data for eQTLs with genome positions.</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_eqtlExp.html">xQTLdownload_eqtlExp()</a></code> </p>
        </td>
        <td><p>Download normalized expression for gene with a variant-gene pair</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_exp.html">xQTLdownload_exp()</a></code> </p>
        </td>
        <td><p>Download normalized gene expression at the sample level for a specified tissue.</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_geneMedExp.html">xQTLdownload_geneMedExp()</a></code> </p>
        </td>
        <td><p>Download median expressions for multiple genes in a specified tissue</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_hqtl.html">xQTLdownload_hqtl()</a></code> </p>
        </td>
        <td><p>Download summary statistics data of H3K4me1 and H3K27ac histone QTL (hQTL) using a specified location</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_hqtlmeta.html">xQTLdownload_hqtlmeta()</a></code> </p>
        </td>
        <td><p>Download metadata for H3K4me1 and H3K27ac histone QTL (hQTL)</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_mQTL.html">xQTLdownload_mQTL()</a></code> </p>
        </td>
        <td><p>Download summary statistics data of DNA methylation QTL (mQTL) using CpG ID</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_mqtlmeta.html">xQTLdownload_mqtlmeta()</a></code> </p>
        </td>
        <td><p>Download metadata of DNA methylation QTL (mQTL)</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_sc.html">xQTLdownload_sc()</a></code> </p>
        </td>
        <td><p>Download all sc-eQTL associations for a specified gene</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_scSig.html">xQTLdownload_scSig()</a></code> </p>
        </td>
        <td><p>Download significant sc-eQTL associations for a specified gene</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_sgene.html">xQTLdownload_sgene()</a></code> </p>
        </td>
        <td><p>Download details of sGenes (sQTL Genes) for a specified gene or a tissue.</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_sqtlAllAsso.html">xQTLdownload_sqtlAllAsso()</a></code> </p>
        </td>
        <td><p>Download summary statistics data for sQTLs with a specified gene or a tissue</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_sqtlExp.html">xQTLdownload_sqtlExp()</a></code> </p>
        </td>
        <td><p>Download normalized PSI value of intron for a sQTL pair</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLdownload_xqtlAllAsso.html">xQTLdownload_xqtlAllAsso()</a></code> </p>
        </td>
        <td><p>Download summary statistics of xQTL for a specified gene, default:3'aQTL</p></td>
      </tr><tr><td>
          <p><code><a href="EBIquery_allTerm.html">EBIquery_allTerm()</a></code> </p>
        </td>
        <td><p>Query supported terms (phenotypes, studies, tissues) in eQTL catalogue</p></td>
      </tr><tr><td>
          <p><code><a href="extractGeneInfo.html">extractGeneInfo()</a></code> </p>
        </td>
        <td><p>Extract gene details from gencodeGeneInfoAllGranges object</p></td>
      </tr><tr><td>
          <p><code><a href="retrieveLD.html">retrieveLD()</a></code> </p>
        </td>
        <td><p>Retrieve SNP pairwise LD from locuscompare database</p></td>
      </tr><tr><td>
          <p><code><a href="retrieveLD_LDproxy.html">retrieveLD_LDproxy()</a></code> </p>
        </td>
        <td><p>Retrieve SNP pairwise LD from LDlink database</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="pre-processing">Pre-processing <a href="#pre-processing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="xQTLanno_calLambda.html">xQTLanno_calLambda()</a></code> </p>
        </td>
        <td><p>Calculate genomic control inflation factor for a QTL/GWAS summary statistics dataset</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLanno_chippeak.html">xQTLanno_chippeak()</a></code> </p>
        </td>
        <td><p>Annotate GWAS / QTL signals using bed4 format table object that provided by user</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLanno_genomic.html">xQTLanno_genomic()</a></code> </p>
        </td>
        <td><p>Genomic annotation of significant signals from GWAS / QTL</p></td>
      </tr><tr><td>
          <p><code><a href="xQTL_export.html">xQTL_export()</a></code> </p>
        </td>
        <td><p>Export expression object to a specified format</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="analyze">Analyze <a href="#analyze" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="xQTLanalyze_coloc.html">xQTLanalyze_coloc()</a></code> </p>
        </td>
        <td><p>Conduct colocalization analysis with trait genes generated from <code>xQTLanalyze_getTraits</code></p></td>
      </tr><tr><td>
          <p><code><a href="xQTLanalyze_coloc_diy.html">xQTLanalyze_coloc_diy()</a></code> </p>
        </td>
        <td><p>Conduct colocalization analysis with customized QTL data</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLanalyze_getSentinelSnp.html">xQTLanalyze_getSentinelSnp()</a></code> </p>
        </td>
        <td><p>Detect sentinel SNPs for GWAS using summary statistics data</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLanalyze_getTraits.html">xQTLanalyze_getTraits()</a></code> </p>
        </td>
        <td><p>Identify trait genes using sentinel SNPs generated from <code>xQTLanalyze_getSentinelSnp</code></p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualize">Visualize <a href="#visualize" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="xQTLvisual_anno.html">xQTLvisual_anno()</a></code> </p>
        </td>
        <td><p>Visualize enrichment of variants derived from <code>xQTLanno_genomic</code></p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_coloc.html">xQTLvisual_coloc()</a></code> </p>
        </td>
        <td><p>Heatmap plot of the LD-p-value relationship of the eQTL</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_eqtl.html">xQTLvisual_eqtl()</a></code> </p>
        </td>
        <td><p>Box plot with jittered points for showing number and significance of eQTL associations</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_eqtlExp.html">xQTLvisual_eqtlExp()</a></code> </p>
        </td>
        <td><p>Boxplot of normalized expression stratified by genotypes for eQTL.</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_geneExpTissues.html">xQTLvisual_geneExpTissues()</a></code> </p>
        </td>
        <td><p>Violin plot of distribution of the gene expression profiles among multiple tissues.</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_genesExp.html">xQTLvisual_genesExp()</a></code> </p>
        </td>
        <td><p>Density plot of expression profiles for multiple genes</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_genoBox.html">xQTLvisual_genoBox()</a></code> </p>
        </td>
        <td><p>Boxplot of values stratified by genotypes with customized data</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_locusCombine.html">xQTLvisual_locusCombine()</a></code> </p>
        </td>
        <td><p>Generate a combined figure including locusZoom and locuscompare plot</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_locusCompare.html">xQTLvisual_locusCompare()</a></code> </p>
        </td>
        <td><p>Dotplot of comparing regional signals between GWAS and xQTL</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_locusZoom.html">xQTLvisual_locusZoom()</a></code> </p>
        </td>
        <td><p>Locuszoom plot for visualizing regional signals relative to genomic position with summary statistics data</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_manhattan.html">xQTLvisual_manhattan()</a></code> </p>
        </td>
        <td><p>Manhattan plot for a GWAS summary statistics dataset</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_PZPlot.html">xQTLvisual_PZPlot()</a></code> </p>
        </td>
        <td><p>Compare P-values reported to P-values calculated from Z statistics derived from the reported beta and standard error.</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_qqPlot.html">xQTLvisual_qqPlot()</a></code> </p>
        </td>
        <td><p>Quantile-quantile plot with p-values from GWAS summary statistics data</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_sqtl.html">xQTLvisual_sqtl()</a></code> </p>
        </td>
        <td><p>Box plot with jittered points for showing number and significance of sQTL associations</p></td>
      </tr><tr><td>
          <p><code><a href="xQTLvisual_sqtlExp.html">xQTLvisual_sqtlExp()</a></code> </p>
        </td>
        <td><p>Boxplot of normalized PSI stratified by genotypes for sQTL.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ruofan Ding, Xudong Zou, Yangmei Qin, Gao Wang, Lei Li.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

