<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Generated a combined plot with two locuszoom plots and a locuscompare
This function is rebuilt from locuscompare.R (https://github.com/boxiangliu/locuscomparer/blob/master/R/locuscompare.R)."><title>xQTLvisual_locusCombine — xQTLvisual_locusCombine • xQTLbiolinks</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="xQTLvisual_locusCombine — xQTLvisual_locusCombine"><meta property="og:description" content="Generated a combined plot with two locuszoom plots and a locuscompare
This function is rebuilt from locuscompare.R (https://github.com/boxiangliu/locuscomparer/blob/master/R/locuscompare.R)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">xQTLbiolinks</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>xQTLvisual_locusCombine</h1>
      
      <div class="d-none name"><code>xQTLvisual_locusCombine.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generated a combined plot with two locuszoom plots and a locuscompare
This function is rebuilt from <code>locuscompare.R</code> (https://github.com/boxiangliu/locuscomparer/blob/master/R/locuscompare.R).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">xQTLvisual_locusCombine</span><span class="op">(</span>
  <span class="va">gwasEqtldata</span>,
  posRange <span class="op">=</span> <span class="st">""</span>,
  population <span class="op">=</span> <span class="st">"EUR"</span>,
  highlightSnp <span class="op">=</span> <span class="st">""</span>,
  legend_position <span class="op">=</span> <span class="st">"bottomright"</span>,
  snpLD <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>gwasEqtldata</dt>
<dd><p>A data.frame or a data.table that including signals from both GWAS and eQTL. Five columns are required (arbitrary column names is supported):
<code>Col 1</code>. "snps" (character), , using an rsID (e.g. "rs11966562").
<code>Col 2</code>. "chromosome" (character), one of the chromosome from chr1-chr22.
<code>Col 3</code>. "postion" (integer), genome position of snp.
<code>Col 4</code>. "P-value" (numeric) of GWAS signals.
<code>Col 5</code>. "P-value" (numeric) of eQTL signals.</p></dd>
<dt>posRange</dt>
<dd><p>Genome range that you want to visualize (e.g. "chr6:3e7-7e7"). Default is the region that covers all snps.</p></dd>
<dt>population</dt>
<dd><p>One of the 5 popuations from 1000 Genomes: 'AFR', 'AMR', 'EAS', 'EUR', and 'SAS'.</p></dd>
<dt>highlightSnp</dt>
<dd><p>Default is the snp that with lowest p-value.</p></dd>
<dt>legend_position</dt>
<dd><p>(string, optional) Either 'bottomright','topright', or 'topleft'. Default: 'bottomright'.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># \donttest{</span></span>
<span class="r-in">  <span class="va">u1</span> <span class="op">&lt;-</span><span class="st">"https://raw.githubusercontent.com/dingruofan/exampleData/master/gwas/AD/gwasEqtldata.txt"</span></span>
<span class="r-in">  <span class="va">gwasEqtldata</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">u1</span><span class="op">)</span></span>
<span class="r-in">  <span class="fu">xQTLvisual_locusCombine</span><span class="op">(</span><span class="va">gwasEqtldata</span>, highlightSnp<span class="op">=</span><span class="st">"rs13120565"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Highlighted SNP: [rs13120565]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying variant... | 2022-6月-01 16:40:33 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Retrieve LD information of SNP: [rs13120565]...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Start plotting locuscomappre...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Start plotting locuszoom for gwas...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Highlighted SNP: [rs13120565]...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Start plotting locuszoom for eQTL... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Highlighted SNP: [rs13120565]...</span>
<span class="r-plt img"><img src="xQTLvisual_locusCombine-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span class="co"># }</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Ruofan Ding.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

