<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="download normalized expression of intron for a sQTL pair."><title>xQTLdownload_sqtlExp — xQTLdownload_sqtlExp • xQTLbiolinks</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="xQTLdownload_sqtlExp — xQTLdownload_sqtlExp"><meta property="og:description" content="download normalized expression of intron for a sQTL pair."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">xQTLbiolinks</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>xQTLdownload_sqtlExp</h1>
      
      <div class="d-none name"><code>xQTLdownload_sqtlExp.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>download normalized expression of intron for a sQTL pair.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">xQTLdownload_sqtlExp</span><span class="op">(</span>
  variantName <span class="op">=</span> <span class="st">""</span>,
  phenotypeId <span class="op">=</span> <span class="st">""</span>,
  variantType <span class="op">=</span> <span class="st">"auto"</span>,
  tissueSiteDetail <span class="op">=</span> <span class="st">""</span>,
  datasetId <span class="op">=</span> <span class="st">"gtex_v8"</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>variantName</dt>
<dd><p>A character string. like dbsnp ID or variant id in GTEx.</p></dd>
<dt>phenotypeId</dt>
<dd><p>A character string. Format like: "chr1:497299:498399:clu_54863:ENSG00000239906.1"</p></dd>
<dt>variantType</dt>
<dd><p>A character string. "auto", "snpId" or "variantId". Default: "auto".</p></dd>
<dt>tissueSiteDetail</dt>
<dd><p>A character string. Tissue detail can be listed using \"tissueSiteDetailGTExv8\" or \"tissueSiteDetailGTExv7\"</p></dd>
<dt>datasetId</dt>
<dd><p>A character string. "gtex_v8" or "gtex_v7". Default: "gtex_v8".</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.table</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># \donttest{</span></span>
<span class="r-in"> <span class="co"># Download exp in different tissues:</span></span>
<span class="r-in"> <span class="fu">xQTLdownload_sqtlExp</span><span class="op">(</span>variantName<span class="op">=</span><span class="st">"rs1450891501"</span>,</span>
<span class="r-in">                      phenotypeId<span class="op">=</span><span class="st">"chr1:497299:498399:clu_54863:ENSG00000239906.1"</span>,</span>
<span class="r-in">                      tissueSiteDetail<span class="op">=</span><span class="st">"Lung"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Check the variant name:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying variant... | 2022-6月-01 16:36:05 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Downloading expression...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> =================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Summary: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ pValue ]:          5.93492283559215e-07</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ pValueThreshold ]: 2.10462e-05</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ nes ]:             1.49360971143302</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ maf ]:             0.00776699029126214</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ error ]:           0.295170581523851</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Normalized expression and genotypes of [515] samples were found for variant [rs1450891501] - phenotypeId [chr1:497299:498399:clu_54863:ENSG00000239906.1] pair in tissue [Lung] in [gtex_v8].</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Genotype: 6 het; 1 hom; 508 ref.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == For more normalization method details, you can visit: https://www.gtexportal.org/home/faq#normalization. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         normExp genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1:  1.5011214         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2:  0.5549300         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3: -0.8925282         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4:  1.6455877         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   5:  1.5287871         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ---                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 511:  0.3806881         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 512: -0.9349318         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 513: -0.7257876         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 514: -1.1425818         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 515:  0.1055783         0</span>
<span class="r-in"></span>
<span class="r-in"> <span class="co"># Dowload exp using variant ID.</span></span>
<span class="r-in"> <span class="fu">xQTLdownload_sqtlExp</span><span class="op">(</span>variantName<span class="op">=</span><span class="st">"chr1_1259424_T_C_b38"</span>,</span>
<span class="r-in">                      phenotypeId<span class="op">=</span><span class="st">" chr1:1487914:1489204:clu_52051:ENSG00000160072.19"</span>,</span>
<span class="r-in">                      tissueSiteDetail<span class="op">=</span><span class="st">"Adipose - Subcutaneous"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Check the variant name:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying variant... | 2022-6月-01 16:36:07 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Downloading expression...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> =================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Summary: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ pValue ]:          2.63274330410663e-06</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ pValueThreshold ]: 6.49936e-06</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ nes ]:             -0.334711684766284</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ maf ]:             0.301204819277108</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [ error ]:           0.0705140406615062</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Normalized expression and genotypes of [581] samples were found for variant [chr1_1259424_T_C_b38] - phenotypeId [ chr1:1487914:1489204:clu_52051:ENSG00000160072.19] pair in tissue [Adipose - Subcutaneous] in [gtex_v8].</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == Genotype: 194 het; 78 hom; 309 ref.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> == For more normalization method details, you can visit: https://www.gtexportal.org/home/faq#normalization. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         normExp genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1:  0.7685717         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2:  1.7084808         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3: -1.1936438         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4: -0.8390463         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   5:  1.5652198         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ---                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 577: -0.0660877         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 578:  0.7391191         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 579:  0.2699149         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 580: -0.2832245         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 581:  0.3222542         0</span>
<span class="r-in"><span class="co"># }</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Ruofan Ding.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

